In [2]:
] add DataStructures

   Updating registry at `C:\Users\Owner\.julia\registries\General`


   Updating git-repo `https://github.com/JuliaRegistries/General.git`


Fetching: [========================================>]  100.0 %

  Resolving package versions...
  Installed Compat ───────────── v3.25.0
  Installed OrderedCollections ─ v1.4.0
  Installed DataStructures ───── v0.18.9
Updating `C:\Users\Owner\.julia\environments\v1.5\Project.toml`
  [864edb3b] + DataStructures v0.18.9
Updating `C:\Users\Owner\.julia\environments\v1.5\Manifest.toml`
  [34da2185] + Compat v3.25.0
  [864edb3b] + DataStructures v0.18.9
  [bac558e1] + OrderedCollections v1.4.0
  [8bb1440f] + DelimitedFiles
  [37e2e46d] + LinearAlgebra
  [1a1011a3] + SharedArrays
  [2f01184e] + SparseArrays
  [10745b16] + Statistics


In [4]:
] add QuadGK

  Resolving package versions...
  Installed QuadGK ─ v2.4.1
Updating `C:\Users\Owner\.julia\environments\v1.5\Project.toml`
  [1fd47b50] + QuadGK v2.4.1
Updating `C:\Users\Owner\.julia\environments\v1.5\Manifest.toml`
  [1fd47b50] + QuadGK v2.4.1


In [6]:
] add Calculus

  Resolving package versions...
  Installed Calculus ─ v0.5.1
Updating `C:\Users\Owner\.julia\environments\v1.5\Project.toml`
  [49dc2e85] + Calculus v0.5.1
Updating `C:\Users\Owner\.julia\environments\v1.5\Manifest.toml`
  [49dc2e85] + Calculus v0.5.1


In [8]:
] add ForwardDiff

  Resolving package versions...
  Installed OpenSpecFun_jll ────────────── v0.5.3+4
  Installed NaNMath ────────────────────── v0.3.5
  Installed DiffResults ────────────────── v1.0.3
  Installed CommonSubexpressions ───────── v0.3.0
  Installed DiffRules ──────────────────── v1.0.2
  Installed CompilerSupportLibraries_jll ─ v0.3.4+0
  Installed ForwardDiff ────────────────── v0.10.16
  Installed StaticArrays ───────────────── v1.0.1
  Installed MacroTools ─────────────────── v0.5.6
  Installed ChainRulesCore ─────────────── v0.9.29
  Installed SpecialFunctions ───────────── v1.3.0
Updating `C:\Users\Owner\.julia\environments\v1.5\Project.toml`
  [f6369f11] + ForwardDiff v0.10.16
Updating `C:\Users\Owner\.julia\environments\v1.5\Manifest.toml`
  [d360d2e6] + ChainRulesCore v0.9.29
  [bbf7d656] + CommonSubexpressions v0.3.0
  [e66e0078] + CompilerSupportLibraries_jll v0.3.4+0
  [163ba53b] + DiffResults v1.0.3
  [b552c78f] + DiffRules v1.0.2
  [f6369f11] + ForwardDiff v0.10.16
  [1914dd2

In [11]:
include("a1x.jl")
include("a2x.jl")
include("a3x.jl");

┌ Info: Precompiling ForwardDiff [f6369f11-7733-5829-9624-2563aa707210]
└ @ Base loading.jl:1278


# LogNormal Variables and Risk Neutral (RN) Valuation

## Change of Density Function

For $X \sim f(x)$ and $g: \mathcal R \rightarrow \mathcal R$, where $\int_{-\infty}^{\infty}|g(x)|dx < \infty$, $Y = g(X)$ is a random variable. Under certain assumption the density of $Y$ can be computed from the density of $X$.

## Lemma 4.1

For $g$ **invertible** and **differentiable**

If $g(x + \delta x) > g(x)$

$$h(y) = \frac{f(g^{-1}(y))}{g'(g^{-1}(y))} \tag{Lemma 4.1}$$

If $g(x + \delta x) < g(x)$

$$h(y) = -\frac{f(g^{-1}(y))}{g'(g^{-1}(y))} \tag{Lemma 4.1}$$

## Note on Computing the Expected Value and Variance of Random Variable $Y = g(x)$

$$
    E[Y] = E[g(X)] = \int_{-\infty}^{\infty}g(x)\ f(x)\ dx \\
    \text{var}(Y) = E[Y^2] - (E[Y])^2 = \int_{-\infty}^{\infty}g^2(x)f(x)\ dx - \left(\int_{-\infty}^{\infty}g(x)f(x)\ dx\right)^2
$$

**Compare this result to Lemma 4.1 and match with example in Lemma 4.3.**

## Lognormal Random Variables

**For Black Scholes derivation**,

$$
    \ln\left(\frac{S(t_2)}{S(t_1)}\right) = \left(\mu-q-\frac{\sigma^2}{2}\right)(t_2-t_1) + \sigma\sqrt{t_2-t_1}Z, \forall\ 0 \leq t_1 < t_2 \leq T
$$

For $X = ln(Y)$ and $Y \sim \text{Lognormal}(\mu, \sigma^2)$,

$$
    \ln(Y) = \mu + \sigma Z
$$

## $\text{PDF}$ of $Y \sim \text{Lognormal}(\mu, \sigma^2)$

## Lemma 4.2

For $Y \sim \text{Lognormal}(\mu, \sigma^2)$

$$
    h(y) = \begin{cases}
        \frac{1}{y|\sigma|\sqrt{2\pi}}\exp(-\frac{(\ln(y)-\mu)^2}{2\sigma^2}), & y \gt 0 \\
        0, & y \leq 0
    \end{cases}
$$

## $E[Y]$ and $var(Y)$ for $Y \sim \text{Lognormal}(\mu, \sigma^2)$

## Lemma 4.3

For $Y \sim \text{Lognormal}(\mu, \sigma^2)$

$$
    E[Y] = \exp\left(\mu + \frac{\sigma^2}{2}\right); \\
    var(Y) = e^{2\mu + \sigma^2}\left(e^{\sigma^2} - 1\right).
$$

## Independent Random Variable

For the following definitions and lemmas, we assume to have the same **probability space**.

## Definitions

$$
    P(\cap_{i = 1}^{n}{\left(a_i\leq X_i \leq b_i\right)}) = \prod_{i=1}^{n}{a_i \leq X_i \leq b_i} \\~\\
    P(a_i \leq X_i \leq b_i | a_j \leq X_j \leq b_j) = P(a_i \leq X_i \leq b_i),\ \forall\ a_i \leq b_i,\ i \in (1,n) \\~\\
    P(A | B) = \frac{P(A \cap B)}{P(B)} \\~\\
    P(\cap_{i=1}^{n}(a_i \leq X_i \leq b_i) = \int_{a_n}^{b_n}\int_{a_{n-1}}^{b_{n-1}}\cdots\int_{a_1}^{b_1}f(x_1,x_2,\ldots,x_n)\ dx_1\ dx_2\ \cdots\ dx_n\ \forall\ a_i \leq b_i
$$

## Lemma 4.5

Random variables $x_1,x_2,\ldots,x_n$ are independent $\iff$ the joint distribution function $f(x_1, x_2, \ldots, x_n)$ of $X_1, X_2, \ldots, X_n$ is

$$
    f(x_1, x_2, \cdots, x_n) = \prod_{i=1}^{n}f(x_i)
$$

## Lemma 4.6

For independent random variables $X_1$, $X_2$, $\cdots$, $X_n$ and continuous functions $h_1:\mathcal R \rightarrow \mathcal R$, $h_2:\mathcal R \rightarrow \mathcal R$, $\cdots$, $h_n: \mathcal R \rightarrow \mathcal R$,

$$
    E[h_1(X_1)h_2(X_2)\cdots h_n(X_n)] = E[h_1(X_1)]\ E[h_2(X_2)]\cdots E[h_n(X_n)]
$$

## Lemma 4.7

For independent random variables $X_1, X_2, \ldots, X_n$,

$$
    cov(X_1, X_2, \ldots, X_n) = corr(X_1, X_2, \ldots, X_n) = 0
$$

and therefore

$$
    var(X_1 + X_2 + \cdots + X_n) = var(X_1) + var(X_2) + \cdots + var(X_n)
$$

## Lemma 4.8

For independent random variables $X_1$, $X_2$, $\ldots$, $X_n$ and continuous functions $h_1:\mathcal R \rightarrow \mathcal R$, $h_2:\mathcal R \rightarrow \mathcal R$, $\ldots$, $h_n: \mathcal R \rightarrow \mathcal R$,

$$
    cov(h_1(X_1), h_2(X_2), \ldots, h_n(X_n)) = corr(h_1(X_1), h_2(X_2), \ldots, h_n(X_n)) = 0
$$

and

$$
    var(h_1(X_1) + h_2(X_2) + \cdots + h_n(X_n)) = var(h_1(X_1)) + var(h_2(X_2)) + \cdots + var(h_n(X_n)).
$$

$\therefore h_1(X_1), h_2(X_2), \ldots, h_n(X_n)$ are independent.

## Theorem 4.1

For independent random variables $X_1$ and $X_2$, with $\text{PDFs}\ f_1(X_1)$ and $f_2(X_2)$, respectively, if $X_1$ and $X_2$ are independent random variables, then the $\text{PDF}$ of the random variable $X_1 + X_2$ is the convolution function,

$$
    (f_1 * f_2)(x) = \int_{-\infty}^{\infty}f_1(z)f_2(x - z)\ dz = \int_{-\infty}^{\infty}f_1(x - z)f_2(z)\ dz
$$

## Independent Normal Random Variables

The sum of two normal random variables is not always random, but when they are *independent*...

## Theorem 4.2

$$
    X_1 + X_2 \sim f(z) = \cal{N}\left(\mu_1 + \mu_2, \sigma_1^2 + \sigma_2^2\right)
$$

where

$$
    f(z) = \frac{1}{\sqrt{2\pi\left(\sigma_1^2 + \sigma_2^2\right)}}\ \exp\left(-\frac{(z - (\mu_1 + \mu_2))^2}{2\left(\sigma_1^2 + \sigma_2^2\right)}\right)
$$

and

$$
    X_1 + X_2 = \mu_1 + \mu_2 + \sqrt{\sigma_1^2 + \sigma_2^2}Z
$$

## Independent Lognormal Random Variables

$$
    Y_1Y_2 \sim \text{Lognormal}\left(\mu_1 + \mu_2, \sqrt{\sigma_1^2 + \sigma_2^2}\right)
$$

where

$$
    \ln(Y_1Y_2) = \ln(Y_1) + \ln(Y_2) = X_1 + X_2 = (\mu_1 + \mu_2) + \sqrt{\sigma_1^2 + \sigma_2^2}Z
$$

## Lognormal evolution of asset prices

Assume spot price, $S(0)$, at time $t = 0$, then

$$
    \frac{S(t + \delta t)-S(t)}{\delta t\ S(t)}\tag{rate of return on $[t, t + \delta t]$}
$$

If $\mu$ exists, then the price S(t) would be an exponential function with

$$
    \frac{S(t + \delta t)-S(t)}{\delta t\ S(t)} = \mu \tag{4.1.1}
$$

$$
    \frac{dS}{dt} = \mu S, \forall\ 0 \lt t \lt T \tag{$t\rightarrow 0$},
$$

$$
    S(t) = S(0)\ e^{\mu t}\tag{solution to ODE above}
$$

$$
    \frac{S(t + \delta t) - S(t)}{S(t)\ \delta t} = \mu + \sigma\frac{1}{\sqrt{\delta t}}Z\Longrightarrow S(t + \delta t) - S(t) = \mu S(t)\delta t + \sigma S(t)\sqrt{\delta t}Z \tag{better model than (1)}
$$

Based on the equation above, which represents the **expected value** ($\mu$) and **standard deviation of asset returns** on the order of $\left(\frac{1}{\sqrt{\delta t}}\right)$, namely $\left(\frac{\sigma}{\sqrt{\delta t}}\right)$, it says on page 124 of the book that the instantaneous price of the underlying asset is required to follow the following stochastic differential equation (SDE):

$$
    dS = \mu Sdt + \sigma SdX
$$

with $t = 0$ given by $S(0)$. Here, $X(t)\ \forall\ t \geq 0$, is a **Weiner process**.

If dividend paying asset

$$
    dS = (\mu - q)Sdt + \sigma SdX
$$

The solution is $S(t)\ \forall\ t \geq 0 \iff S(t)$ satisfies the following lognormal model:

$$
    \ln{\left(\frac{S(t_2)}{S(t_1)}\right)} = \left(\mu - q - \frac{\sigma^2}{2}\right)(t_2 - t_1) + \sigma\sqrt{t_2 - t_1}Z, \forall\ 0 \leq t_1 \leq t_2 \tag{4.1.2}
$$

where $Z$ is the standard normal variable.

## Risk-neutral derivation of the Black-Scholes formula

A derivative security with payoff $V(T)$ at maturity $T$ given by risk-neutral valuation has value

$$
    V(0) = e^{rT}\ E_{RN}[V(T)]\tag{4.1.3}
$$

where the expected value is computed with respect to the value of $S(T)$ given by $(2)$ for $\mu = r$, $t_1 = 0$, $t_2 = T$, i.e.,

$$
    \ln{\left(\frac{S(T)}{S(0)}\right)} = \left(r - q - \frac{\sigma^2}{2}\right)T + \sigma\sqrt{T}Z
$$

which is equivalent to 

$$
    S(T) = S(0)\ \exp\left({\left(r - q - \frac{\sigma^2}{2}\right)T + \sigma\sqrt{T}Z}\right)\tag{4.1.4}
$$

Equation (3) is used to compute risk-neutral pricing for European Call and Put options, as below,

$$
    C(0) = e^{rT}\ E_{RN}[\max(S(T) - K, 0)]\\
    P(0) = e^{rT}\ E_{RN}[\max(K - S(T), 0)]
$$

where the expected value is computed with respect to $S(T)$ given by (4).

By definition

$$
    C(T) = \max(S(T) - K, 0) =
    \begin{cases}
        S(T) - K & S(T) \geq K \\
        0 & S(T) \lt K
    \end{cases}
    \\~\\
    P(T) = \max(K - S(T), 0) =
    \begin{cases}
        0 & S(T) \gt K \\
        K - S(T) & S(T) \leq K
    \end{cases}
$$

From (4), it follows that

$$
    S(T) \geq K \iff Z \geq \frac{\ln{\left(\frac{K}{S(0)}\right)}-\left(r - q - \frac{\sigma^2}{2}\right)\ T}{\sigma\sqrt{T}} = -d_2
$$

where

$$
    d_2 = \frac{\ln\left(\frac{S(0)}{K}\right) + \left(r - q - \frac{\sigma^2}{2}\right)\ T}{\sigma\sqrt{T}}
$$

$\therefore$

$$
    C(T) = max(S(T) - K, 0) = \begin{cases}
        S(0)\ \exp{\left[\left(r - q - \frac{\sigma^2}{2}\right)T+\sigma\sqrt{T}Z\right]} - K & Z \geq -d_2 \\
        0 & Z \lt -d_2
    \end{cases}
    \\~\\
    P(T) = max(K - S(T), 0) = \begin{cases}
        0 & Z \gt -d_2 \\
        K - S(0)\ \exp{\left[\left(r - q - \frac{\sigma^2}{2}\right)T+\sigma\sqrt{T}Z\right]} & Z \leq -d_2
    \end{cases}
$$

An argument is given, starting on pp. $125-126$, proving that the value of C(0) is as follows (worth noting the exercize that follows):

$$
    C(0) = S(0)e^{-qT}N(d_1) - Ke^{-rT}N(d_2)
$$

## Probability that an European option expires in the money (ITM)

For a **call option**, the probability of expiring in the money is as follows:

$$
    P(S(T) > K) = P\left(\frac{S(T)}{S(0)} \gt \frac{K}{S(0)}\right) = P\left(\ln{\left(\frac{S(T)}{S(0)}\right)} \gt \ln{\left(\frac{K}{S(0)}\right)}\right) \\
    \ \ \ \ \ \ \ \ \ \ \ = P\left(\left(\mu - q - \frac{\sigma^2}{2}\right)T + \sigma\sqrt{T}Z \gt \ln{\left(\frac{K}{S(0)}\right)}\right) \\
    = P\left(Z \gt \frac{\ln{\left(\frac{K}{S(0)}\right)} - \left(\mu - q - \frac{\sigma^2}{2}\right)T}{\sigma\sqrt{T}}\right) \ \ \\
    = P(Z > a) \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \
$$

where

$$
    a = \frac{\ln{\left(\frac{K}{S(0)}\right)} - \left(\mu - q - \frac{\sigma^2}{2}\right)T}{\sigma\sqrt{T}}
$$

**Discussion continues on p. $129$, which shows that**

$$
    P(S(T) \gt K) = N(d_{2,\mu}) = N\ \left(\frac{\ln{\left(\frac{S(0)}{K}\right)} + \left(\mu - \frac{\sigma^2}{2}\right)T}{\sigma\sqrt{T}}\right) \tag{for call option}
$$

For a **put option**, the probability of expiring in the money is then,

$$
    P(S(T) < K) = 1 - P(S(T) > K) = 1 - N(d_{2,\mu}) = N(-d_{2,\mu}).\tag{for put option}
$$

In practice, use algorithm `a2x.cum_dist_normal` to approximate the two probabilities above. Also, remember that $\mu = r \rightarrow d_{2,\mu} = d_{2}$ from `a2x.d2`.

## Financial interpretation of $N(d_1)$ and $N(d_2)$ from the Black-Scholes formulas

To reiterate from earlier discussions, the Black-Scholes formula for a **call option** is

$$
    C = S(0)e^{-q(T -t)}N(d_1) - Ke^{-r(T -t)}N(d_2),
$$

and the **Delta of a call option** is $\Delta{C} = e^{-q(T - t)}N(d_1)$. For $q = 0$,

$$
    N(d_1) = \Delta(C),
$$

which is not true for dividend paying assets, since in that case $\Delta{C} = e^{-qT}N(d_1)$.

The Black-Scholes formula for the **put option** is

$$
    P = Ke^{-r(T - t}N(-d_2) - Se^{-q(T - t)}N(-d_1).
$$

Remember that the **Delta of a put option** is $\Delta{P} = -e^{-q(T - t)}N(-d_1)$. Then, for q = 0,

$$
    \Delta{P} = -N(-d_1).
$$

The term $N(-d_2)$ represents the risk-neutral probability that a **put option** expires ITM. Remember,

$$
    \text{put option expires ITM} \iff \text{call option with same strike and maturity expires OTM}.
$$

## $\S 4.2$ Exercises

**Problem 11**: If the underlying asset of a plain vanilla European put option follows a lognormal distribution, the value $S(T)$ of the underlying asset at maturity is a lognormal variable given by

$$
    \ln(S(T)) = \ln(S(0)) + \left(r - q - \frac{\sigma}{2}\right)T + \sigma\sqrt{T}Z \tag{4.2.1}
$$

where σ is the volatility of the underlying asset. Then, the probability density function $h(y)$ of $S(T)$ is

$$
    h(y) = \frac{1}{yσ\sqrt{2\pi T}}\exp{\left( -\frac{\left(\ln y - \ln(S(0)) - \left(r - q - \frac{\sigma^2}{2}\right)T\right)^2}{2\sigma^2T} \right)}
$$

if $y \gt 0$, and $h(y) = 0$ if $y \leq 0$.

Using risk-neutral valuation, we find that the value of the put is given by

$$
    P = e^{-rT}E_{RN}[\max(K - S(T), 0)] \tag{4.2.2}
$$

$$
    = e^{-rT}\int_{0}^{K}(K - y)h(y)\ dy \ \ \
$$

where $h(y)$ is given by (1).

The Black-Scholes value of the put is $P_{BS} = 4.863603$. To compute the numerical approximation of the integral (2), we start with a partition of the interval $[0, K]$ into $4$ intervals, and double the numbers of intervals up to 8192 intervals. We report the Midpoint Rule and Simpson's Rule approximation to (2) and the corresponding approximation errors to the Black-Scholes value $P_{BS}$.

In [6]:
using DataFrames

t = 0
S = exp(1)
K = 50
T = 0.5
σ = 1
r = 0.04
q = 0.02

nd1(s) = -1 ⋅ d1(t, s, K, T, σ, r, q)
nd2(s) = -1 ⋅ d2(t, s, K, T, σ, r, q)

N(z) = 1 / sqrt(2*π) ⋅ exp(-(z ^ 2 / 2))

# h(y) = 1 / (y*σ*\sqrt(2πT))
# f(x) = exp(-rT)*(K - y)h(y)
# numericals = Array{Float64, 2}(undef, 15, 4)
numericals = zeros(15, 4)
no_intervals, I_midpoint = I_numerical(0, 2, N(nd1), Midpoint, 1e-6)
no_intervals, I_midpoint = I_numerical(0, 2, N(nd2), Midpoint, 1e-6)

no_intervals, I_simpson = I_numerical(0, 2, N(nd1), Simpson, 1e-6)
no_intervals, I_simpson = I_numerical(0, 2, N(nd1), Simpson, 1e-6)

no_intervals = [4 ⋅ 2^i for i in 0:max(size(I_midpoint)[1], size(I_simpson)[1])-1]

for i in 1:max(size(I_midpoint)[1], size(I_simpson)[1])
    numericals[i, 1] = no_intervals[i]
end

for i in 1:size(I_midpoint)[1]
    numericals[i, 2] = I_midpoint[i]
end

for i in 1:size(I_simpson)[1]
    numericals[i, 3] = I_simpson[i]
end

# Pbs = K ⋅ exp(-r(T - t)) ⋅ N(nd2) - S ⋅ exp(-q(T - t)) ⋅ N(nd1)b

df = DataFrame(Intervals = no_intervals, Midpoint = numericals[:, 2], Simpson = numericals[:, 4])

MethodError: MethodError: objects of type Float64 are not callable

## Note: Problem #11 in Ch 4 Needs Reviewing

# Nonlinear Solvers, N-dimensional Newton's Method, Implied Volatility & Bootstrapping

Various methods are discussed for solving $f(x) = 0$, where $f(x)$ is nonlinear.

There are several examples of nonlinear problems in financial mathematics, such as:

* Computing the yield of a bond;
* Computing implied volatility;
* Using bootstrapping to find zero rate curves from bond prices.

## Three methods for solving one dimensional nonlinear problems: Bisection Method, Secant Method and Newton's Method.

## Bisection Method

**This method works best for small intervals**.

This is the simplest method, where $f$ is a function $f:[a, b] \rightarrow \cal R$ such that **$\textbf{f(a)}$ and $\textbf{f(b)}$ have different signs**. From the **intermediate value theorem**, **$\textbf{f(x) = 0}$ for at least one point**, say $x$, on $[a, b]$. The bisection method will find only one (1) trivial solution.

Interesting to note: the bisection method reminds me of the **divide and conquer** algorithm I learned in computer science, where we split the interval $[a, b]$, find a center point, $c = \frac{a + b}{2}$ and keep searching for a value, in this case, the $f(x) = 0$ point.

The search is for different signs, e.g., f(a) and f(c) have different signs, or f(c) and f(b) have signs. Once the "sign $\Delta$" is found, we search in *that* smaller interval and discard the other interval, e.g., make $c' = \frac{a + c}{2}$ or $c' = \frac{c + b}{2}$, so that the search narrows to $[a,c]$ or $[c, b]$ with a division at c'. Repeat the algorithm while $f(x) = 0$ is not found within the given tolerance factor.

In [289]:
function bisection(a, b, f; tol_int = 10e-6, tol_approx = 10e-9)
    if (f(a)f(b) >= 0)
        return -1
    end
    xL = a; xR = b
    xM = (xL + xR) / 2
    while( (max(abs(f(xL)), abs(f(xR))) > tol_approx) || (xR - xL > tol_int) )
        xM = (xL + xR) / 2
        if (f(xL)f(xM) < 0)
            xR = xM          # active interval [xL, xM]
        else
            xL = xM          # active interval [xM, xR]
        end
    end
    return xM
end

bisection (generic function with 1 method)

## Theorem 5.1

Based on the **Mean Value Theorem**, this states that if $f: [a, b] \rightarrow \cal R$ is a continuous function and $f(a)$ and $f(b)$ have opposite signs, then the bisection method applied on the interval $[a, b]$ converges to a solution of $f(x) = 0$.

To show this result mathematically, note that must be a constant $C \gt 0$, $\forall\ x \in [a, b]$, such that

$$
    |f'(x)| \leq C, \forall\ x \in [a, b]
$$

and from the Mean Value Theorem:

$$
    \left|\frac{f(x_R) - f(x_L)}{x_R - x_L}\right| = f'(\xi),
$$

so that

$$
    |f(x_R) - f(x_L)| \leq C\ |x_R - x_L| = C\frac{b - a}{2^n}.
$$

In the text, its also mentioned that for a $n$ large enough, the bisection method will always converge, since

$$
    \max{f(x_L), f(x_R)} \leq |f(x_R) - f(x_L)| \leq C\frac{b - a}{2^n}.
$$

$\therefore$ provided some agreeable stopping condition, the bisection method will always converge $\iff$ $f(a)$ and $f(b)$ have opposite signs.

**Example**: Using the bisection method on the interval [-2, 3], to find the zero of $f(x) = 0$,

$$
    f(x) = x^4 - 5x^2 + 4 - \frac{1}{1 + e^{x^3}}
$$

In [20]:
f(x) = x^4 - 5x^2 + 4 - 1 / (1 + exp(x^3))

bisection(-2, 3, f)

-0.8896418153308332

## Newton's Method

Where $x_k$ is the approximation of an exact solution $x^*$ to $f(x) = 0$, after $k$ iterations, a tangent line to $f(x)$ at the point $(x_k, f(x_k))$ is approximately

$$
    x_{k+1} = x_{k} - \frac{f(x_k)}{f'(x_k)},\ \forall\ k \geq 0. \tag{7}
$$

It is worth noting an argument for the Taylor expansion of $f(x)$ around the point $x_{k}$, presented on pp. $138-139$ in the text, which states that $f(x_{k + 1}) \approx f(x_k) + (x - x_k)f'(x_k)$, followed by another argument regarding the **bond yield**, which is always positive ($x_0 = 0.1$ is a good guess for the **yield**, $x_0 = 0.2$ for **implied volatility**).

## Theorem 5.2

Where $x^*$ is a solution to $f(x) = 0$, $f$ being twice differentiable, with $f''(x)$ continuous, if $f'(x^*) \neq 0$ and $x_0$ is close enough to $x^*$, then the Newton's method converges quadratically (there is $M \gt 0$ and $n_M$ positive), such that

$$
    \frac{|x_{k + 1} - x^{*}|}{|x_k - x^{*}|^2} = \left|\frac{f''(c_k)}{2f'(x_k)}\right| < M,\ \forall\ k \geq n_M.
$$

In [30]:
] add Calculus

   Updating registry at `~/.julia/registries/General`


   Updating git-repo `https://github.com/JuliaRegistries/General.git`


Fetching: [========================================>]  100.0 %

  Resolving package versions...
  Installed Calculus ─ v0.5.1
   Updating `~/.julia/environments/v1.5/Project.toml`
   49dc2e85 + Calculus v0.5.1
   Updating `~/.julia/environments/v1.5/Manifest.toml`
   49dc2e85 + Calculus v0.5.1


In [52]:
using Calculus

function newton(x0, f; tol_approx = 10e-9, tol_consec = 10e-6)
    n = 0 # number of iterations
    xnew = x0; xold = x0 - 1
    while ( abs(f(xnew)) > tol_approx || abs(xnew - xold) > tol_consec )
        n = n + 1
        xold = xnew
        xnew = xold - f(xold) / derivative(f, xold)
    end
    return n, xnew
end

newton (generic function with 1 method)

**Example**: Let $f(x) = x^4 - 5x^2 + 4 - \frac{1}{1 + e^{x^3}}$. Using Newton's Method to solve $f(x) = 0$, for $x$:

In [36]:
using DataFrames

n = zeros(4)
x = zeros(4)
f(x) = x^4 - 5x^2 + 4 - 1 / (1 + exp(x^3))

n[1], x[1] = newton(-3, f)
n[2], x[2] = newton(-0.5, f)
n[3], x[3] = newton(0.5, f)
n[4], x[4] = newton(3, f)

df = DataFrame(Iteration = n, Solution = x)

┌ Info: Precompiling DataFrames [a93c6f00-e57d-5684-b7b6-d8193f3e46c0]
└ @ Base loading.jl:1276


4×2 DataFrame
│ Row │ Iteration │ Solution  │
│     │ Float64   │ Float64   │
├─────┼───────────┼───────────┤
│ 1   │ 6.0       │ -2.0743   │
│ 2   │ 4.0       │ -0.889642 │
│ 3   │ 4.0       │ 0.950748  │
│ 4   │ 6.0       │ 2.00003   │

## Secant Method

Sometimes, a closed formula for $f(x)$, and therefore, $f'(x)$, is unknown. A **first order finite difference** approximation can be used to estimate the term $f'(x_k)$ from Newton's method in (7). This would be as follows:

$$
    f'(x_k) \approx \frac{f(x_k) - f(x_{k - 1})}{x_k - x_{k - 1}};
$$

for $a = x_k$, $h = x_k - x_{k - 1}$ and $a - h = x_{k - 1}$ in $f(a) \approx \frac{f(a) - f(a - h)}{h}$. The recursion for the secant method is

$$
    x_{k + 1} = x_k - \frac{(x_k - x_{k - 1})f(x_k)}{f(x_k) - f(x_{k - 1})}, \forall\ k \geq 0.
$$

In [353]:
function secant(xm1, x0, f; tol_approx = 10e-9, tol_consec = 10e-6)
    n = 0 # number of iterations
    xnew = x0; xold = xm1
    while ( abs(f(xnew)) > tol_approx || abs(xnew - xold) > tol_consec )
        n = n + 1
        xoldest = xold
        xold = xnew
        xnew = xold - f(xold) ⋅ ((xold - xoldest) / (f(xold) - f(xoldest)))
    end
    return n, xnew
end

secant (generic function with 1 method)

**Example**: similar to the previous examples, we are solving for x:

$$
    f(x) = x^4 - 5x^2 + 4 - \frac{1}{1 + e^{x^3}}
$$

In [35]:
using DataFrames

f(x) = x^4 - 5x^2 + 4 - 1 / (1 + exp(x^3))
x0 = -3
n1, x1 = secant(x0 - 0.01, x0, f)
n2, x2 = newton(x0, f)

df = DataFrame(SecantN = n1, SecantX = x1, NewtonN = n2, NewtonX = x2)

1×4 DataFrame
│ Row │ SecantN │ SecantX │ NewtonN │ NewtonX │
│     │ Int64   │ Float64 │ Int64   │ Float64 │
├─────┼─────────┼─────────┼─────────┼─────────┤
│ 1   │ 8       │ -2.0743 │ 6       │ -2.0743 │

As per above, Newton's method merges faster to the real value of $x$. **One must use Newton's method, unless one is absolutely sure that there is no closed formula for $x$, in which case, the Secant method would work as intended.**

## Numerical methods for N-dimensional nonlinear problems

Here, there are no practical anologs of the bisection method, but Newton's method can be extended to $N$ dimensional nonlinear problems; also, too can be extended the approximate Newton's method, using **finite difference approximations**.

## The N-dimensional Newton's Method

Details are given in the book (pp. $144 - 145$), illustrating how to compute the gradient $DF(x)$ of $F(x)$, 

$$
    DF(x) = \left(\begin{matrix}
            \frac{\partial{F_1}}{\partial{x_1}}(x) & \frac{\partial{F_1}}{\partial{x_2}}(x) & \cdots & \frac{\partial{F_1}}{\partial{x_N}}(x) \\
            \frac{\partial{F_2}}{\partial{x_1}}(x) & \frac{\partial{F_2}}{\partial{x_2}}(x) & \cdots & \frac{\partial{F_2}}{\partial{x_N}}(x) \\
            \vdots & \vdots & \vdots & \vdots \\
            \frac{\partial{F_N}}{\partial{x_1}}(x) & \frac{\partial{F_N}}{\partial{x_2}}(x) & \cdots & \frac{\partial{F_N}}{\partial{x_N}}(x)
            \end{matrix}\right),
$$

to solve the nonlinear problem

$$
    F(x) = 0.
$$

The recursion for Newton's method follows from the linear Taylor expansion of $F(x)$.

$$
    F(x) \approx F(x_k) + DF(x_k)(x - x_k)
$$

The random variable $x_{x + 1}$ (where $x_{k + 1}$ is substituted for $x$) is shown to be

$$
    x_{k + 1} = x_{k} - (DF(x_k))^{-1}F(x_k),\ \forall\ k \geq 0,
$$

but since this is computationally expensive, there has to be another way to solve the nonlinear equation $F(x) = 0$; in fact, computing the vector $v_k = DF(x_k))^{-1}F(x_k)$ is equivalent to solving the linear system

$$
    DF(x_k)v_k = F(x_k).
$$

In [2]:
] add ForwardDiff

   Updating registry at `~/.julia/registries/General`


   Updating git-repo `https://github.com/JuliaRegistries/General.git`


    Fetching: [=======>                                 ]  16.0 %Fetching: [========================================>]  100.0 %===================>                    ]  49.8 %65.8 %Fetching: [====================================>    ]  88.1 %

  Resolving package versions...
  Installed DiffResults ────────── v1.0.2
  Installed NaNMath ────────────── v0.3.4
  Installed DiffRules ──────────── v1.0.1
  Installed CommonSubexpressions ─ v0.3.0
  Installed StaticArrays ───────── v0.12.4
  Installed ForwardDiff ────────── v0.10.12
   Updating `~/.julia/environments/v1.5/Project.toml`
   f6369f11 + ForwardDiff v0.10.12
   Updating `~/.julia/environments/v1.5/Manifest.toml`
   bbf7d656 + CommonSubexpressions v0.3.0
   163ba53b + DiffResults v1.0.2
   b552c78f + DiffRules v1.0.1
   f6369f11 + ForwardDiff v0.10.12
   77ba4419 + NaNMath v0.3.4
   90137ffa + StaticArrays v0.12.4


In [4]:
using ForwardDiff, LinearAlgebra

function nDimNewton(x0, F; tol_approx = 10e-9, tol_consec = 10e-6)
    t1 = [tol_approx for I in 1:size(F)[1]]; tol_approx = t1
    n = 0
    xnew = x0; xold = x0 - ones(size(x0)[1])
    
    while ( [norm(F[I](xnew)) for I in 1:size(F)[1]] > tol_approx  || norm(xnew - xold) > tol_consec )
        n = n + 1
        xold = xnew
        DF = zeros(size(F)[1], size(F)[1])
        for I in 1:size(F)[1]
            DF[I, :] = ForwardDiff.gradient(F[I], xold)
        end
        xnew = xold - DF \ [F[I](xold) for I in 1:size(F)[1]]
    end
    return (n, xnew)
end

nDimNewton (generic function with 1 method)

In [63]:
using DataFrames

f(x) = x[1]^3 + 2x[1] ⋅ x[2] + x[3]^2 - x[2] ⋅ x[3] + 9
g(x) = 2x[1]^2 + 2x[1] ⋅ x[2]^2 + x[2]^3 ⋅ x[3]^2 - x[2]^2 ⋅ x[3] - 2
h(x) = x[1] ⋅ x[2] ⋅ x[3] + x[1]^3 - x[3]^2 - x[1] ⋅ x[2]^2 - 4

funcs = Function[]
push!(funcs, f)
push!(funcs, g)
push!(funcs, h)

c = Channel(41)
function FuncCh()
    foreach(i->put!(c, funcs[i]), 1:size(funcs)[1])
end

FuncCh()

CAP = 3
output = Function[]

while(size(output)[1] < CAP)
    push!(output, take!(c))
end

close(c)
F = output

x0 = [1, 2, 3]
n, x = nDimNewton(x0, F)
df = DataFrame(iterations = n, valuesOfX = x)

3×2 DataFrame
│ Row │ iterations │ valuesOfX │
│     │ Int64      │ Float64   │
├─────┼────────────┼───────────┤
│ 1   │ 9          │ -1.69055  │
│ 2   │ 9          │ 1.98311   │
│ 3   │ 9          │ -0.884558 │

In [3]:
x0 = [2, 2, 2]
n, x = nDimNewton(x0, F)
df = DataFrame(iterations = n, valuesOfX = x)

3×2 DataFrame
│ Row │ iterations │ valuesOfX │
│     │ Int64      │ Float64   │
├─────┼────────────┼───────────┤
│ 1   │ 40         │ -1.0      │
│ 2   │ 40         │ 3.0       │
│ 3   │ 40         │ 1.0       │

## Theorem 5.3

Where $x^*$ is a solution to $F(x) = 0$, with $F$ having continuous $2^{nd}$ order partial derivatives, if $DF(x^*)$ is a nonsingular matrix and $x_0$ is close enough to $x^*$, then Newton's method converges quadratically; there exists $M > 0$ and $n_M$
positive, such that
    
$$
    \frac{||x_{k + 1} - x^*||}{||x_k - x^*||^2} < M,\ \forall k \geq n_M.
$$

## The Approximate Newton's Method

In cases where a closed formula becomes impossible or inefficient to find, one can use **finite differences approximations** as weapons of last resort, as illustrated by the following equation, where:

$$
    \frac{\partial F_i}{\partial x_j} \approx \Delta_j F_i(x) = \frac{F_i(x + he_j) - F_i(x)}{h},
$$

for the $(i, j)$th position in the DF(x) approximate matrix,

$$
    \Delta F(x) = \left(\begin{matrix}
            \Delta_1F_1(x) & \Delta_2F_1(x) & \cdots & \Delta_NF_1(x) \\
            \Delta_1F_2(x) & \Delta_2F_2(x) & \cdots & \Delta_NF_2(x) \\
            \vdots & \vdots & \vdots & \vdots \\
            \Delta_1F_N(x) & \Delta_2F_N(x) & \cdots & \Delta_NF_N(x)
            \end{matrix}\right).
$$

One basically ends up with the following recursion for the approximate Newton's method.

$$
    x_{k + 1} = x_k - (\Delta F(x_k))^{-1}F(x_k)
$$

In [128]:
using ForwardDiff, LinearAlgebra
    
function approxNewton(x0, F; tol_approx = 10e-9, tol_consec = 10e-6)
    t1 = [tol_approx for I in 1:size(F)[1]]; tol_approx = t1
    n = 0
    h = tol_consec; xnew = x0; xold = x0 - ones(size(x0)[1])
    
    while ( [norm(F[I](xnew)) for I in 1:size(F)[1]] > tol_approx  || norm(xnew - xold) > tol_consec )
        n = n + 1
        xold = xnew
        ΔF = zeros(size(F)[1], size(F)[1])
        # e = [K == J ? 1 : 0 for K in 1:size(F)[1], J in 1:size(F)[1]]
        for I in 1:size(F)[1]
            for J in 1:size(F)[1]
                e = [K == J ? 1 : 0 for K in 1:size(F)[1]]
                ΔF[I, J] = (F[I](xold .+ h .⋅ e) - F[I](xold)) / h
            end
        end
        xnew = xold - ΔF \ [F[I](xold) for I in 1:size(F)[1]]
    end
    return (n, xnew)
end

approxNewton (generic function with 2 methods)

In [129]:
x0 = [1, 2, 3]
n, x = approxNewton(x0, F)
df = DataFrame(iterations = n, valuesOfX = x)

3×2 DataFrame
│ Row │ iterations │ valuesOfX │
│     │ Int64      │ Float64   │
├─────┼────────────┼───────────┤
│ 1   │ 9          │ -1.69055  │
│ 2   │ 9          │ 1.98311   │
│ 3   │ 9          │ -0.884558 │

In [130]:
x0 = [2, 2, 2]
n, x = approxNewton(x0, F)
df = DataFrame(iterations = n, valuesOfX = x)

3×2 DataFrame
│ Row │ iterations │ valuesOfX │
│     │ Int64      │ Float64   │
├─────┼────────────┼───────────┤
│ 1   │ 37         │ -1.0      │
│ 2   │ 37         │ 3.0       │
│ 3   │ 37         │ 1.0       │

There is also **central finite differences** (cf. pp. $147-148$ of text); done in the exercises below. Note that central finite differences approximates the gradient more closely that if forard finite differences are used, resulting in an iteration count closer to Newton's method.

## Financial Applications

## Computing bond yields

If $c_i,\ i = 1:n$ is the cash payments received by the holder of a bond with face value $B$ and $t_i$ is the time when payment $c_i$ is made, then the yield of the bond $y$ is the interest rate of return of the bond, i.e.,

$$
    B = \sum_{i = 1}^{n}c_i\exp{(-yt_i)}
$$

If the price of the bond $B$ and all future cash flows $c_i$ are known, this then becomes the following nonlinear problem

$$
    f(x) = 0, \tag{8}
$$

which by setting $x = y$, the solution becomes

$$
    f(x) = \sum_{i = 1}^{n}c_ie^{xt_i} - B \approx 0. \tag{9}
$$

Solving (8) using Newton's method, it is easy to see that

$$
    f'(x) = -\sum_{i = 1}^{n}t_ic_ie^{-xt_i}. \tag{10}
$$

Using $(9)$ and $(10)$, the recursion formula $(7)$ for solving $(8)$ becomes

$$
    x_{k + 1} = x_k - \frac{f(x_k)}{f'(x_k)} = x_k + \frac{\sum_{i = 1}^nc_ie^{-x_kt_i - B}}{\sum_{i = 1}^{n}t_ic_ie^{-x_kt_i}}.
$$

**A good initial guess for the bond yield is $10\%$.**

In [70]:
function bond_yield(B, n, t_cash_flow, v_cash_flow; tol = 10e-6)
    x0 = 0.1
    xnew = x0; xold = x0 - 1
    while (abs(xnew - xold) > tol)
        xold = xnew
        xnew = xold + (sum(v_cash_flow[1:n] .* exp.(-xold .* t_cash_flow[1:n])) - B) / sum(t_cash_flow[1:n] .* v_cash_flow[1:n] .* exp.(-xold .* t_cash_flow[1:n]))
    end
    return xnew
end

bond_yield (generic function with 1 method)

**Example:** A 34 months semiannual coupon bond with coupon rate $8%$ and face value $100$ is priced at $105$. The yield of the bond is

In [145]:
t_cash_flow = [4 / 12, 10 / 12, 16 / 12, 22 / 12, 28 / 12, 34 / 12]
v_cash_flow = [4.0, 4.0, 4.0, 4.0, 4.0, 104.0]
bond_yield(105, 6, t_cash_flow, v_cash_flow)

0.06450193139556404

## Computing implied volatility

Using the same $C$, $P$, $S$, $K$ and $T$ values as in $(46)$ and $(47)$ of $\cal{FE}1$ (assuming t = 0), we solve

$$
    f(x) = 0, \tag{11}
$$

for

$$
    f(x) = Se^{qT}N(d_1(x)) - Ke^{-rT}N(d_2(x)) - C,
$$

when $f(x)$ represents a call option and

$$
    f(x) = Ke^{rT}N(-d_2(x)) - Se^{-qT}N(-d_1(x)) - P,
$$

when $f(x)$ represents a put option. Here,

$$
    d_1(x) = \frac{\ln{\left(\frac{S}{K}\right)} + \left(r - q + \frac{x^2}{2}\right)}{x\sqrt{T}};\ 
    d_2(x) = \frac{\ln{\left(\frac{S}{K}\right)} + \left(r - q - \frac{x^2}{2}\right)}{x\sqrt{T}}
$$

The value of $\sigma$ thus computed is the implied volatility corresponding to a given option price. Since the aim is to solve ${11}$ using Newton's method, one must remember that $f'(x)$ is the vega of a plain vanilla European put or call option, where

$$
    f'(x) = \frac{1}{2\pi}Se^{-qT}\sqrt{T}\exp\left(-\frac{(d_1(x))^2}{2}\right).
$$

Since $f'(x) \geq 0\ \forall\ x$, it follows that the function $f(x)$ is strictly increasing, and therefore $(11)$ has **at most** one solution. The recursion formula $(7)$ for solving $(11)$ is then

$$
    x_{k + 1} = x_k - \frac{f(x_k)}{f'(x_k)}.
$$

**A good initial guess is $25\%$ volatility ($x_0 = 0.25$).**

In [265]:
using Calculus

function implied_volatility(C, S, K, T, q, r; tol = 10e-6, x0 = 0.25)
    f_BS(x) = S * exp(-q * T)cum_dist_normal(d1x(S, K, T, x, r, q)) - K * exp(-r * T)cum_dist_normal(d2x(S, K, T, x, r, q)) - C
    vega_BS(x) = derivative(f_BS, x)
    xnew = x0; xold = x0 - 1
    while (abs(xnew - xold) > tol)
        xold = xnew
        xnew = xold - f_BS(xold) / vega_BS(xold)
    end
    return xnew
end

implied_volatility (generic function with 1 method)

In [268]:
implied_volatility(7, 25, 20, 1, 0, 0.05)

0.36306322295713117

## Bootstrapping for finding zero rate curves

The zero rate $r(0, t)$ between time $0$ and time $t$ is the rate of return of a cash deposit made at time $0$ and maturing at time $t$, for continuously compounded interest rates. The price $B$ of a bond with future cash flows $c_i$ to be paid to the holder of a bond on dates $t_i$, $i = 1:n$, is given by
    
$$
    B = \sum_{i = 1}^{n}c_ie^{-r(0, t_i)t_i} \tag{12}
$$

If the prices of serveral different bonds are known, then there is information on the zero rates corresponding to each date when a bond makes a payment. Since it is discrete and finite, the general form of the zero rate curve $r(0,t)$ for all times $t$ smaller than the largest maturity of any of the bonds cannot be determined uniquely.

The bootstrapping method provides a simplified framework in which the zero rate curve can be uniquely determined froma given set of bond prices (assuming zero credit risk, i.e., that the bonds never default), by requiring that the zero rate $r(0,t)$ is linear between any two consecutive bond maturities.

**Example**: The prices and coupon rates of four semiannual coupon bonds with face value $100$ are as follows:

$$
    \begin{matrix} \text{Maturity} & \text{Coupon Rate} & {Price} \\
                   \text{6 months} & 0 & 99 \\
                   \text{1 year} & 4 & 102 \\
                   \text{2 year} & 4 & 103.5 \\
                   \text{5 years} & 4 & 109
                   \end{matrix}
$$

Bootstrapping is used to fit a zero rate interest curve $r(0,t), 0 \leq t \leq 5$ to the prices of these bonds, by assuming that the zero rate curve is linear between every two consecutive maturities of the bonds. Expression (11) can be used to obtain the following formula for the price of a semiannual coupon bond

$$
    B = \sum_{i = 1}^{n - 1}\frac{C}{2}\ 100\ e^{-r(0,t_i)t_i} + \left( 100 + \frac{C}{2}\ 100\right)\ e^{-r(0,T)T} \tag{13}
$$

where $C$ is the coupon rate, $T$ is the maturity of the bond, and $t_i, i = 1:(n-1)$, are all the coupon dates before maturity. The solution is as follows:

For the six month

In [6]:
e = exp(1)
f(x) = 100e^(-0.5x) - 99
x0 = 0.01; n, r05 = newton(x0, f)

(3, 0.020100671707002797)

For the one year

In [9]:
f(x) = 2e^(-0.5r05) + 102e^(-x) - 102
x0 = 0.01; n, r1 = newton(x0, f)

(3, 0.0196026472935135)

For the two year

In [10]:
f(x) = 2e^(-0.5r05) + 2e^(-r1) + 2e^(-1.5(r1 + x) / 2) + 102e^(-2x) - 103.5
x0 = 0.01; n, r2 = newton(x0, f)

(3, 0.021944274303852037)

For $\frac{1}{2}$ a year, we take an average, since the price at maturity $1.5$ is unknown.

In [11]:
r15 = (r1 + r2) / 2

0.020773460798682766

For the five year bond, pricing formula $(13)$ becomes

$$
    109 = 2\sum_{i = 1}^9e^{-0.5i\ r(0, 0.5i)} + 102\ e^{-5r(0,5)} \tag{14}
$$

Assuming $r(0,t)$ linear for $t \in [2, 5]$, i.e.,

$$
    \begin{matrix}
        r(0, t) & = & \frac{f - t}{|f - s|}r(0,s) + \frac{t - s}{|f - s|}r(0,f), \ \forall\ s\leq t \leq f \\
        r(0, t) & = & \frac{5 - t}{3}r(0,2) + \frac{t - 2}{3}r(0,5),\ \forall\ 2\leq t \leq 5
    \end{matrix}
$$

then letting $x = r(0,5)$ with

\begin{matrix}
    r(0,2.5) & = & \frac{x + 5r(0,2)}{6} \\
    r(0, 3) & = & \frac{x + 2r(0,2)}{3} \\
    r(0, 3.5) & = & \frac{x + r(0,2)}{2} \\
    r(0, 4) & = & \frac{2x + r(0,2)}{3} \\
    r(0, 4.5) & = & \frac{5x + r(0,2)}{6}
\end{matrix}

formula $(14)$ becomes the following non-linear equation which can be solved using Newton's method.

In [12]:
f(x) = 2e^(-0.5r05) + 2e^(-r1) + 2e^(-1.5(r1 + r2) / 2) + 2e^(-2r2) + 2e^(-2.5((x + 5r2) / 6)) + 2e^(-3((x + 2r2) / 3)) + 2e^(-3.5((x + r2) / 2)) + 2e^(-4((2x + r2) / 3))+ 2e^(-4.5((5x + r2) / 6)) + 102e^(-5x) - 109
x0 = 0.01; n, r5 = newton(x0, f)

(3, 0.02080186020158749)

$\therefore\text{ the following zero rates corresponding to the maturities of the four given bonds were found}$

$$
    \begin{matrix}
        r(0, 0.5) & = & 0.0201007 & = & 2.0101\% \\
        r(0, 1) & = & 0.0196026 & = & 1.9603\% \\
        r(0, 2) & = & 2.1944 & = & 2.1944\% \\
        r(0, 5) & = & 0.0208019 & = & 2.0802\%
    \end{matrix}
$$

Given the overnight lending rate $r(0, 0)$, $\forall t \in [0.5, 5]$

$$
    \begin{matrix}
        2t \cdot r(0, 0.5) & + & (1 - 2t)\ r(0, 0) & \hspace{2em} & \text{if t} \in [0.0, 0.5] \\
        (2t - 1) \cdot r(0, 1.0) & + & 2(1 - t)\ r(0, 0.5) & \hspace{2em} & \text{if t} \in [0.5, 1.0] \\
        (t - 1) \cdot r(0, 2.0) & + & (2 - t)\ r(0, 1) & \hspace{2em} & \text{if t} \in [1.0, 2.0] \\
        \frac{t - 2}{3} \cdot r(0, 5.0) & + & \frac{5 - t}{3}\ r(0, 2) & \hspace{2em} & \text{if t} \in [2.0, 5.0]
    \end{matrix}
$$

**Pain Points**

I tried doing the following, but **Julia** does not allow certain uses of operators, like `*`, `^` and `+` would not work. If one is to modify the output of a function on the fly, one would have to modify the original function itself.

In [14]:
coeff = [0.5I for I in 1:9]; R = Function[];

# known
r05 = r05
r10 = r10
r15= (r1 + r2) / 2
r2 = r2; RK = r05 + r1 + r15 + r2

# unknown
r25(x) = (x + 5r2) / 6
r3(x) = (x + 2r2) / 3
r35(x) = (x + r2) / 2
r4(x) = (2x + r2) / 3
r45(x) = (5x + r2) / 6
# r50(x) = 102e^(-5x) - 109

# push!(R, r05); push!(R, r10); push!(R, r15); push!(R, r20)
push!(R, r25); push!(R, r3); push!(R, r35); push!(R, r4); push!(R, r45)

f(x) = 2 .* ( sum([e^(-coeff[I]R[I-4]) for I in 5:9]) + r05 + r1 + r15 + r2 ) + 102e^(-5x) - 109

# f(x) = 2e^(-0.5r05) + 2e^(-r1) + 2e^(-1.5(r1 + r2) / 2) + 2e^(-2r2) + 2e^(-2.5((x + 5r2) / 6)) + 2e^(-3((x + 2r2) / 3)) + 2e^(-3.5((x + r2) / 2)) + 2e^(-4((2x + r2) / 3))+ 2e^(-4.5((5x + r2) / 6)) + 102e^(-5x) - 109
x0 = 0.01; n, r5 = newton(x0, f)

LoadError: MethodError: no method matching *(::Float64, ::typeof(r25))
Closest candidates are:
  *(::Any, ::Any, !Matched::Any, !Matched::Any...) at operators.jl:540
  *(::Union{Float16, Float32, Float64}, !Matched::BigFloat) at mpfr.jl:393
  *(::AbstractFloat, !Matched::Bool) at bool.jl:108
  ...

# $\S 5.2$ Exercises

**Problem 1**: Using Newton's method to find the yield of a five year semiannual coupon bond with $3.375\%$ coupon rate and price $100\frac{1}{32}$, the duration and convexity of the bond are...

**From Theory**:

$$
    B = \sum_{i = 1}^{n - 1}c_i \ \exp\left(-y\ \frac{i}{K}\right) \ +\ (100 + c_n)\exp{(-yT)},\hspace{2em} \text{where}\ c_i =  \frac{r}{K}\ \forall\ i \leq n, K \in {2, 4, ...}
$$

As per the usual, although the bond price is $100\frac{1}{32}$, a face value of $\$100$ is used here. Nine ($5 \times 2 - 1$) $\$\left(100 \times \frac{0.03375}{2}\right) = \$1.6875$ coupon payments are made every six months, and a final payment of $\$\left(100 + \frac{0.03375}{2}\right) = \$101.6875$ is made after five years. By writing the value of the bond in terms of its yield...

$$
    100 + \frac{1}{32} = B = \sum_{i = 1}^{n}1.6875\ \exp{\left(-y\ \frac{i}{2}\right)}\ +\ 101.6875\ \exp(-5y) \hspace{16em}\hspace{1ex}
$$

becomes the nonlinear equation to be solved; with initial guess $x_0 = 0.1$,

In [277]:
B = 100.03125; n = 10
t_cash_flow = [0.5, 1, 1.5, 2, 2.5, 3, 3.5, 4, 4.5, 5]
v_cash_flow = [1.6875 for I in 1:9]; append!(v_cash_flow, 100 + 1.6875)
y = bond_yield(B, n, t_cash_flow, v_cash_flow)

0.03340109853803073

To compute the bond duration and convexity, I borrow a function called `bond_PrDuCv` from $\cal FE1$, the $\S$ on **bond duration & convexity**, which I included in `a2x.jl`. I must make note of the fact that the duration and convexity of the bond are given by

$$
    D = \frac{1}{B}\left(\sum_{i = 1}^{9}1.6875\frac{i}{2}\exp\left(-y\ \frac{i}{2}\right) + 101.6875 \cdot 5\exp(-5y)\right) \\
    C = \frac{1}{B}\left(\sum_{i = 1}^{9}1.6875\frac{i^2}{4}\exp\left(-y\ \frac{i}{2}\right) + 101.6875 \cdot 25\exp(-5y)\right),
$$

where $B = 100 + \frac{1}{32}$ and $y \approx 0.033401$. The duration and convexity then is...

In [278]:
B, D, C = bond_PrDuCv(5, 10, t_cash_flow, v_cash_flow, y)

(100.03125000009307, 4.642735048425559, 22.573118015851666)

**Problem 2:** To compute the price, yield, modified duration and convexity of a two year semiannual coupon bond with face value 100 and coupon rate 8% (coupon rate given by $r(0, t) = 0.05 + 0.01\ln (1 + \frac{t}{2})$, I borrow functions called `get_cash_flow` and `bond_price_zero_rate`, from $\cal FE1$ $\S$ on **zero rate** and **instantaneous rate** curves.

In [68]:
Y = 2; m = 2; T = 12 * m; F = 100; C = 0.08; n = convert(Int64, Y / 0.5); t_cash_flow, v_cash_flow = get_cash_flow(n, Y, T, F, C)

([0.5, 1.0, 1.5, 2.0], [4.0, 4.0, 4.0, 104.0])

In [69]:
f(t) = 0.05 + 0.01 * log(1 + t / 2)
disc, B = bond_price_zero_rate(4, t_cash_flow, v_cash_flow, f)

([0.9742223482612451, 0.9473803297277678, 0.919988382423771, 0.8923802539163926], 104.17391064895598)

**Problem 4**: Given a three month ATM call (S = K) on an underlying asset with spot price 30 paying dividends continuously at a $2\%$ rate, worth $\$2.5$ with constant risk-free rate of $6\%$, (i) computing the implied volatility on $[0.0001, 1]$ using the bisection method, secant method with initial guess $0.5$ and Newton's method with initial guess $0.5$ and (ii) using the formula

$$
    \sigma_{imp,approx} \approx \frac{\sqrt{2\pi}}{S\sqrt{T}}\ \frac{C - \frac{(r - q)T}{2}S}{1 - \frac{(r + q)T}{2}}
$$

to compute an approximate value $\sigma_{imp_approx}$ for the implied volatility, also computing the relative error $\frac{|\sigma_{imp,approx} - \sigma_{imp}|}{\sigma_{imp}}.$

In [396]:
C = 2.5; S = 30; K = 30; T = 3 / 12; r = 0.06; q = 0.02
f(x) = S * exp(-q * T)cum_dist_normal(d1x(S, K, T, x, r, q)) - (K * exp(-r * T)cum_dist_normal(d2x(S, K, T, x, r, q))) - C
bisection(0.0001, 1, f)

0.39704818831067523

In [397]:
n, x = secant(0.6, 0.5, f; tol_approx = 10e-9, tol_consec = 10e-6)

(3, 0.39704818680444287)

In [398]:
n, x = newton(0.5, f)

(3, 0.397048186803561)

They like to use this formula for computing an approximate value for the implied volatility

$$
    \sigma_{imp} \approx \frac{\sqrt{2\pi}}{S\sqrt{T}}\frac{C - \frac{(r - q)T}{2}S}{1 - \frac{(r + q)T}{2}},
$$

which gives a value of

In [399]:
σ = √(2π) / (S * √(T)) * (C - (((r - q) * T) / 2) * S) / (1 - (r + q)T / 2)

0.39667181450389566

The relative error $\frac{|\sigma_{imp,approx} - \sigma_{imp}|}{\sigma_{imp}}$ then is

In [402]:
abs(σ - x) / σ

0.000948825416638382

or $0.0948\%$.

**Problem 5**: The same as done above for `nDimNewton` except that we have the **central difference approximation** replace ***forward finite differences***. Where before the **approximate gradient** was $\Delta_j F_i(x) = \frac{F_i(x + he_j) - F_i(x)}{h}$, with central differences, it becomes $\Delta_j F_i(x) = \frac{F_i(x + he_j) - F_i(x) - he_j}{2h}$.

In [27]:
using ForwardDiff, LinearAlgebra
    
function approxNewton(x0, F; tol_approx = 10e-9, tol_consec = 10e-6)
    t1 = [tol_approx for I in 1:size(F)[1]]; tol_approx = t1
    n = 0
    h = tol_consec; xnew = x0; xold = x0 - ones(size(x0)[1])
    
    while ( [norm(F[I](xnew)) for I in 1:size(F)[1]] > tol_approx  || norm(xnew - xold) > tol_consec )
        n = n + 1
        xold = xnew
        ΔF = zeros(size(F)[1], size(F)[1])
        # e = [K == J ? 1 : 0 for K in 1:size(F)[1], J in 1:size(F)[1]]
        for I in 1:size(F)[1]
            for J in 1:size(F)[1]
                e = [K == J ? 1 : 0 for K in 1:size(F)[1]]
                ΔF[I, J] = (F[I](xold .+ h .* e) - F[I](xold .- h .* e)) / 2h
            end
        end
        xnew = xold - ΔF \ [F[I](xold) for I in 1:size(F)[1]]
    end
    return (n, xnew)
end

approxNewton (generic function with 1 method)

In [5]:
using DataFrames

f(x) = x[1]^3 + 2x[1] ⋅ x[2] + x[3]^2 - x[2] ⋅ x[3] + 9
g(x) = 2x[1]^2 + 2x[1] ⋅ x[2]^2 + x[2]^3 ⋅ x[3]^2 - x[2]^2 ⋅ x[3] - 2
h(x) = x[1] ⋅ x[2] ⋅ x[3] + x[1]^3 - x[3]^2 - x[1] ⋅ x[2]^2 - 4

funcs = Function[]
push!(funcs, f)
push!(funcs, g)
push!(funcs, h)

c = Channel(41)
function FuncCh()
    foreach(i->put!(c, funcs[i]), 1:size(funcs)[1])
end

FuncCh()

CAP = 3
output = Function[]

while(size(output)[1] < CAP)
    push!(output, take!(c))
end

close(c)
F = output

x0 = [1, 2, 3]
n, x = approxNewton(x0, F)
df = DataFrame(iterations = n, valuesOfX = x)

3×2 DataFrame
│ Row │ iterations │ valuesOfX │
│     │ Int64      │ Float64   │
├─────┼────────────┼───────────┤
│ 1   │ 9          │ -1.69055  │
│ 2   │ 9          │ 1.98311   │
│ 3   │ 9          │ -0.884558 │

In [6]:
x0 = [2, 2, 2]
n, x = approxNewton(x0, F)
df = DataFrame(iterations = n, valuesOfX = x)

3×2 DataFrame
│ Row │ iterations │ valuesOfX │
│     │ Int64      │ Float64   │
├─────┼────────────┼───────────┤
│ 1   │ 40         │ -1.0      │
│ 2   │ 40         │ 3.0       │
│ 3   │ 40         │ 1.0       │

As mentioned above, the **central difference approximation** has a lower iteration count, in particular for $x_0 = [2, 2, 2]$ ($40$ vs $65$ with **forward finite differences**.

**Problem 6**: Solving for $f(x) = 0$ in

$$
    f(x) = e^{xT}\ N\left(\frac{(r - x)\sqrt{T}}{\sigma} + \frac{\sigma\sqrt{T}}{2}\right) - 0.5 \lt 0.085 = r + \frac{\sigma^2}{2}
$$

the following value of q is found in three (3) iterations.

In [25]:
T = 6 / 12; r = 0.04; σ = 0.30; x0 = 0.10
f(x) = exp(-x * T)cum_dist_normal((r - x) * √(T) / σ + σ * √(T) / 2) - 0.5
n, q = newton(x0, f)

(3, 0.06690563571881601)

Worthwhile reading discussion on pp. $133- 134$ on solutions handbook. Also worthwhile reading problems $7-9$ on **bootstrapping**.

**Problem 7**: (i) Using bootstrapping to obtain a zero rate curve from the following prices of Treasury instruments with semiannual coupon payments:

$$
    \begin{matrix} \text{_} & \text{Coupon Rate} & \text{Price} \\
                   \text{3 --- Month T-bill} & 0 & 98.7 \\
                   \text{6 --- Month T-bill} & 0 & 97.5 \\
                   \text{2 --- Month T-bill} & 4.875 & 100\frac{5}{32} \\
                   \text{3 --- Month T-bill} & 4.875 & 97.5\frac{5}{32} \\
                   \text{5 --- Month T-bill} & 4.625 & 97.5\frac{22}{32} \\
                   \text{10 --- Month T-bill} & 4.875 & 97.5\frac{4}{32}\end{matrix}
$$

Assume continuously compounded interest.

(ii) Want to know how the zero rate curves obtained by bootstrapping from the bond prices above, one corresponding to the semi-annually compounded interest, and the other one corresponding to continuously computed interest, compare, i.e., will one of the two curves by higher or lower than the other one, and why?

**Solution**: (i) For the Treasury bills, the zero rate can be computed directly:

Recalling that $r(0, t) = \frac{1}{t}\int_0^t r(\tau)\ d\tau$,

$$
    r(0, 0.25) = 4 \ln \left(\frac{100}{98.7}\right) = 0.052341 = 5.2341\% \\
    r(0, 0.5) = 2 \ln \left(\frac{100}{97.5}\right) = 0.050636 = 5.0636\%
$$

because $\int_0^t \frac{1}{x^\tau}\ d\tau \approx \ln(x^\tau) \approx \tau\ \ln(x)$. Using code:

In [1]:
r(p, t) = log(p^(1 / t))
mo = [3, 6]; p = [100 / 98.7, 100 / 97.5]; t = [mo[1] / 12, mo[2] / 12]
rp = [r(p[1], t[1]), r(p[2], t[2])]

2-element Vector{Float64}:
 0.0523409581946222
 0.05063561596857951

Bootstrapping is needed to obtain the $2$-year, $3$-year, $5$-year and $10$-year zero rates. For example, for the two year bond, if the zero rate curve is assumed to be linear between six months and two years, then
        
$$
    r(0, t) = \frac{(2 - t)r(0, 0.5) + (t - 0.5)r(0, 2)}{1.5}{1.5}, \forall\ t \in [0.5, 2]
$$

If we let $x = r(0, 2); s = 0.5, f = 2$, then assuming $t \in [0.5, 2]$ we find from
$$
    \begin{matrix}
        r(0, t) & = & \frac{f - t}{|f - s|}r(0,s) + \frac{t - s}{|f - s|}r(0,f),\ \forall\ s\leq t \leq f \\
        r(0, t) & = & \frac{(2 - t)r(0, 0.5)\ +\ (t - 0.5)r(0, 2)}{1.5}
    \end{matrix}
$$

that

$$
    r(0, 1) = \frac{r(0, 0.5) + 0.5x}{1.5};\ r(0, 1.5) = \frac{0.5r(0, 0.5) + x}{1.5}
$$

Recalling that the price of the two year bond is the discounted present value of all the future cash flows of the bond,

$$
    100 + \frac{5}{32} = \left(\begin{matrix} 2.4375\ e^{-0.5r(0,0.5)} + 2.4375\ e^{-r(0,1)}
                        + 2.4375\ e^{-1.5r(0,1.5)} + 102.4375\ e^{-2r(0,2)} \\
                        2.4375\ e^{-0.5r(0,0.5)} + 2.4375\ \exp\left({-\frac{r(0,0.5) + 0.5x}{1.5}}\right)
                        + 2.4375\ \exp\left({-1.5\frac{0.5r(0,0.5) + x}{1.5}}\right) + 102.4375\ e^{-2x} \end{matrix}\right)
$$

and using Newton's method to solve the nonlinear equation for $x$,

In [3]:
e = exp(1); 
r05 = rp[2]
f(x) = 2.4375e^(-0.5*r05) + 
       2.4375e^(-(r05 + 0.5x) / 1.5) + 
       2.4375e^(-1.5 * ((0.5 * r05 + x) / 1.5)) + 102.4375e^(-2x) - (100 + 5 /32)

f (generic function with 1 method)

In [4]:
x0 = 0.01; n, r2 = newton(x0, f)

(3, 0.04728926180475103)

Hence, $r(0, 2) = 4.7289\%$.

For $t \in [0.5, 2]$

$$
    r(0, t) = \frac{(2 - t)r(0,0.5) + (t - 0.5)r(0,2)}{1.5}
$$

Assuming that the zero rate curve is linear between two years and three years, $r(0, 0.5)$, $r(0, 1)$, $r(0, 1.5)$, and $r(0, 2)$ are known. If we let $x = r(0, 3)$, the price of the three year bond can be written as

In [5]:
f(x) = (r05 + 0.5x) / 1.5

f (generic function with 1 method)

In [6]:
r1 = f(r2)

0.04952016458063668

In [7]:
f(x) = (0.5r05 + x) / 1.5

f (generic function with 1 method)

In [8]:
r15 = f(r2)

0.048404713192693856

**The following is assuming $t \in [2, 3]$**, where

$$
    r(0, t) = \frac{(3 - t)r(0, 2) + (t - 2)r(0, 3)}{3 - 2}, \hspace{2em} \forall\ 2 \leq t \leq 3
$$

In [9]:
e = exp(1); f(x) =  2.4375e^(-0.5r05) + 
                    2.4375e^(-r1) + 
                    2.4375e^(-1.5r15) + 
                    2.4375e^(-2r2) + 
                    2.4375e^(-2.5((r2 + x) / 2)) + 
                    102.4375e^(-3x) - (100 + 5 /32)

f (generic function with 1 method)

In [10]:
x0 = 0.01; n, r3 = newton(x0, f)

(4, 0.04758187545229992)

Therefore, $r(0, 3) = 4.7582\%$. Something similar can be done to obtain the rates for the five (5) year and $10$ year semi-annual coupon securities.

For the five year zero rate curve, the assumption would be that $t$ is linear for $3 \leq t \leq 5$, so that

$$
    r(0, t) = \frac{(5 - t)r(0, 3) + (t - 3)r(0, 5)}{5 - 3}, \hspace{2em} \forall\ 3 \leq t \leq 5
$$

and

$$
    \begin{matrix}
        r(0, 3.5) & = & \frac{3r(0, 3) + r(0, 5)}{4} \\
        r(0, 4) & = & \frac{r(0, 3) + r(0, 5)}{2} \\
        r(0, 4.5) & = & \frac{r(0, 3) + 3r(0, 5)}{4}
    \end{matrix}
$$

For the ten year zero rate curve, the assumption would be that $t$ is linear for $5 \leq t \leq 10$, so that

$$
    r(0, t) = \frac{(10 - t)r(0, 5) + (t - 5)r(0, 10)}{10 - 5}, \hspace{2em} \forall\ 5 \leq t \leq 10
$$

and

$$
    \begin{matrix}
        r(0, 5.5) & = & \frac{9r5 + r10}{10} \\
        r(0, 6) & = & \frac{4r5 + r10}{5} \\
        r(0, 6.5) & = & \frac{7r5 + 3r10}{10} \\
        r(0, 7) & = & \frac{3r5 + 2r10}{5} \\
        r(0, 7.5) & = & \frac{r5 + r10}{2} \\
        r(0, 8) & = & \frac{2r5 + 3r10}{5} \\
        r(0, 8.5) & = & \frac{3r5 + 7r10}{10} \\
        r(0, 9) & = & \frac{r5 + 4r10}{5} \\
        r(0, 9.5) & = & \frac{r5 + 9r10}{10}
    \end{matrix}
$$

Finally, with semi-annual coupon rate $2.3125\%$,

In [11]:
e = exp(1); f(x) =  2.3125e^(-0.5r05) + 
                    2.3125e^(-r1) + 
                    2.3125e^(-1.5r15) + 
                    2.3125e^(-2r2) + 
                    2.3125e^(-2.5(r2 + r3) / 2) + 
                    2.3125e^(-3r3) + 
                    2.3125e^(-3.5(3r3 + x) / 4) +
                    2.3125e^(-4(r3 + x) / 2) + 
                    2.3125e^(-4.5(r3 + 3x) / 4) +
                    102.3125e^(-5x) - (99 + 22 / 32)

x0 = 0.01; n, r5 = newton(x0, f)

(4, 0.04630335938077369)

and, with semi-annual coupon rate $2.4375\%$,

In [12]:
e = exp(1); f(x) =  2.4375e^(-0.5r05) + 
                    2.4375e^(-r1) + 
                    2.4375e^(-1.5r15) + 
                    2.4375e^(-2r2) + 
                    2.4375e^(-2.5(r2 + r3) / 2) + 
                    2.4375e^(-3r3) + 
                    2.4375e^(-3.5(3r3 + r5) / 4) +
                    2.4375e^(-4(r3 + r5) / 2) +
                    2.4375e^(-4.5(r3 + 3r5) / 4) +
                    2.4375e^(-5r5) +
                    2.4375e^(-5.5(9r5 + x) / 10) +
                    2.4375e^(-6(4r5 + x) / 5) +
                    2.4375e^(-6.5(7r5 + 3x) / 10) +
                    2.4375e^(-7(3r5 + 2x) / 5) +
                    2.4375e^(-7.5(r5 + x) / 2) +
                    2.4375e^(-8(2r5 + 3x) / 5) +
                    2.4375e^(-8.5(3r5 + 7x) / 10) +
                    2.4375e^(-9(r5 + 4x) / 5) +
                    2.4375e^(-9.5(r5 + 9x) / 10) +
                    102.4375e^(-10x) - (101 + 4 / 32)

x0 = 0.01; n, r10 = newton(x0, f)

(4, 0.046772065321197015)

**Problem 8**: Use bootstrapping to obtain a continuously compounded zero rate curve given the prices of the following semiannual coupon bonds:

$$
    \begin{matrix} \text{Maturity} & \text{Coupon Rate} & \text{Price} \\
                   \text{6 months} & 0 & 97.5 \\
                   \text{1 year} & 5 & 100 \\
                   \text{20 months} & 6 & 103 \\
                   \text{40 months} & 5 & 102 \\
                   \text{5 year} & 4 & 103 \end{matrix}
$$

Assume that the overnight rate is $5\%$ and that the zero rate curve is linear on the following time intervals:

$$
    \begin{matrix} [0, 0.5]; & [0.5,1]; & \left[1, \frac{5}{3}\right] & \left[\frac{5}{3}, \frac{10}{3}\right]; & \left[\frac{10}{3}, 5\right]
    \end{matrix}
$$

**Solution**: the overnight rate, $r(0, 0) = 0.05$.

It is quite simple getting the $6$-month zero rate, which is

In [4]:
r05 = 2 * log(100 / 97.5)

0.05063561596857957

or $5.0636\%$

Given the price of the one year T-note, again I use Newton's method to solve for the one year coupon bond.

In [5]:
e = exp(1)
x0 = 0.01
f(x) = 2.5e^(-0.5r05) + 102.5e^(-x) - 100

n, r1 = newton(x0, f)

(3, 0.049369600302769254)

The third bond is a $20$-month security. Since payments are semiannual, the bond pays every

In [63]:
get_periods(20, 6)

Queue{Int64}(Deque [[20, 14, 8, 2]])

six months, so the coupon payment arrives in $2$, $8$, $14$ and $20$ months, also paying the face value at $20$.

Then,

$$
    103 = 3\ \exp\left( -\frac{2}{12}r\left( 0, \frac{2}{12} \right) \right) + 3\ \exp\left( -\frac{8}{12}r\left( 0, \frac{8}{12} \right) \right) + 3\ \exp\left(-\frac{14}{12}r\left( 0, \frac{14}{12} \right) \right) + 103\ \exp\left( -\frac{20}{12}r\left( 0, \frac{20}{12} \right) \right)
$$

I follow the same equation for **linear interpolation**:
    
$$
    r(0, t) = \frac{(f - t)r(0,s) + (t - s)r(0,f)}{|f - s|}, \ \forall\ s\leq t \leq f
$$

then, assuming linear $r\left(0, \frac{2}{12}\right) \in \{[0, 0.5], [0.5, 1]\}$

In [14]:
r01667 = (4 * 0.05 + 2r05) / 6

0.05021187198952653

and for $r\left(0, \frac{8}{12}\right)$

In [15]:
r06667 = (4 * r05 + 2r1) / 6

0.0502136107466428

So then, if I let $x = r\left( 0, \frac{20}{12} \right)$, I take note of the assumption that $r\left(0, \frac{14}{12}\right)$ is linear on $\left[1, \frac{20}{12}\right]$ and therefore

$r\left(0, \frac{14}{12}\right) = \frac{20 / 12 - 14 / 12}{8 / 12}r\left(0,\frac{12}{12}\right) + \frac{14 / 12 - 12 / 12}{8 / 12}r\left(0,\frac{20}{12}\right) = \frac{6r(0, 1) + 2x}{8}$

Following the formula (and I got an error calling `(14/12)(6r1 + 2x) / 8` because it was thinking of `14/12` as a function)

In [16]:
x0 = 0.01
f(x) = 3 * exp(-(2/12)r01667) + 
       3 * exp(-(8/12)r06667) + 
       3 * exp(-(14/12) * (6r1 + 2x)/8) + 
       103 * exp(-(20/12)x) - 103

n, r16667 = newton(x0, f)

(3, 0.05298265375710815)

**For $r(0, t) > 20$ months** with linear interpolation

$$
    r(0, t) = \frac{f - t}{|f - s|}r(0,s) + \frac{t - s}{|f - s|}r(0,f), \ \forall\ s\leq t \leq f
$$

In [2]:
get_periods(40, 6)

Stack{Int64}(Deque [[40, 34, 28, 22, 16, 10, 4]])

I first interpolate the $4$, $10$ and $16$ month zero rates, which are (using the correct coupon interval here):

For $4$ months, with linear $r\left(0, \frac{4}{12}\right) \in \left[0, \frac{1}{2}\right]$

In [10]:
r03333 = (1 / 3) * 0.05 + (2 / 3) * r05

0.05042374397905304

For $10$ months, with linear $r\left(0, \frac{10}{12}\right) \in \left[\frac{1}{2}, 1\right]$

In [11]:
r08333 = (1 / 3) * r05 + (2 / 3) * r1

0.049791605524706024

For $16$ months, with linear $r\left(0, \frac{16}{12}\right) \in \left[1, \frac{5}{3}\right]$

In [17]:
r13333 = (1 / 2) * r1 + (1 / 2) * r16667

0.051176127029938706

Then, to get zero rates for $22$, $28$, $34$ and $40$ months, I first establish $x = r\left(0, \frac{10}{3}\right)$, and using linear interpolation for $t \in \left[\frac{5}{3}, \frac{10}{3}\right]$:

$$
    \begin{matrix}
        r\left(0, \frac{22}{12}\right) & = & \frac{9}{10} \cdot r\left(0, \frac{5}{3} \right) + \frac{1}{10}x \\
        r\left(0, \frac{28}{12}\right) & = & \frac{3}{5} \cdot r\left(0, \frac{5}{3} \right) + \frac{2}{5}x \\
        r\left(0, \frac{34}{12}\right) & = & \frac{3}{10} \cdot r\left(0, \frac{5}{3} \right) + \frac{7}{10}x
    \end{matrix}
$$

then using bootstrapping and Newton's method for $x = r\left(0, \frac{10}{3}\right)$:

In [19]:
x0 = 0.01
f(x) = (5 / 2) * exp(-(4/12)r03333)  + 
       (5 / 2) * exp(-(10/12)r08333) + 
       (5 / 2) * exp(-(16/12)r13333) + 
       (5 / 2) * exp(-(22/12) * ((9 / 10)r16667 + (1 / 10)x)) + 
       (5 / 2) * exp(-(28/12) * ((3 / 5)r16667 + (2 / 5)x)) + 
       (5 / 2) * exp(-(34/12) * ((3 / 10)r16667 + (7/ 10)x)) +
       (100 + (5 / 2)) * exp(-(40/12)x) - 102

n, r33333 = newton(x0, f)

(4, 0.045325682607213195)

**For $r(0, t) > 40$ months** with linear interpolation

$$
    r(0, t) = \frac{f - t}{|f - s|}r(0,s) + \frac{t - s}{|f - s|}r(0,f), \ \forall\ s\leq t \leq f
$$

For $x = r(0, 5)$:

In [22]:
get_periods(60, 6)

Stack{Int64}(Deque [[60, 54, 48, 42, 36, 30, 24, 18, 12, 6]])

I have to determine $t \in \left[\frac{10}{3}, 5\right]$, but first I interpolate for $t \lt \frac{10}{3}$:

For $6$ month with linear $r\ (0, t) \in \left[0, \frac{1}{2}\right]$

In [6]:
r05

0.05063561596857957

For $12$ month with linear $r\ (0, t) \in \left[\frac{1}{2}, 1\right]$

In [7]:
r1

0.049369600302769254

For $18$ months with linear $r(0, t) \in \left[1, \frac{5}{3}\right]$

$r\left(0, \frac{3}{2}\right) = \frac{1}{4} \cdot r\ (0, 1) + \frac{3}{4} \cdot r\left(0, \frac{5}{3}\right)$

In [21]:
r15 = (1 / 4) * r1 + (3 / 4) * r16667

0.052079390393523425

For $24$, $30$ and $36$ months with linear $r(0, t) \in \left[\frac{5}{3}, \frac{10}{3}\right]$

$$
    \begin{matrix}
        r(0, 2) & = & \frac{4}{5} \cdot r\left(0, \frac{5}{3}\right) + \frac{1}{5} \cdot r\left(0, \frac{10}{3}\right) \\
        r\left(0, \frac{5}{2}\right) & = & \frac{1}{2} \cdot r\left(0, \frac{5}{3}\right) + \frac{1}{2} \cdot r\left(0, \frac{10}{3}\right) \\
        r(0, 3) & = &  \frac{1}{5} \cdot r\left(0, \frac{5}{3}\right) + \frac{4}{5} \cdot r\left(0, \frac{10}{3}\right)
    \end{matrix}
$$

which amount to

In [20]:
r2 = (4 / 5) * r16667 + (1 / 5) * r33333
r25 = (1 / 2) * r16667 + (1 / 2) * r33333
r3 = (1 / 5) * r16667 + (4 / 5) * r33333
[r2, r25, r3]

3-element Vector{Float64}:
 0.051451259527129166
 0.049154168182160674
 0.04685707683719219

Then for $42$, $48$ and $54$ months with linear $r(0, t) \in \left[\frac{10}{3}, 5\right]$:

$$
    \begin{matrix}
        r\left(0, \frac{7}{2}\right) & = & \frac{9}{10} \cdot r\left(0, \frac{10}{3}\right) + \frac{1}{10} \cdot r\ (0, 5) \\
        r\ (0, 4) & = & \frac{3}{5} \cdot r\left(0, \frac{10}{3}\right) + \frac{2}{5} \cdot r\ (0, 5) \\
        r\left(0, \frac{9}{2}\right) & = & \frac{3}{10} \cdot r\left(0, \frac{10}{3}\right) + \frac{7}{10} \cdot r\ (0, 5)
    \end{matrix}
$$

Finally, using bootstrapping and the Newton-Raphson method 

In [23]:
x = 0.01
f(x) = 2 * exp(-(6/12)r05) +
       2 * exp(-(12/12)r1) +
       2 * exp(-(18/12)r15) + 
       2 * exp(-(24/12)r2) +
       2 * exp(-(30/12)r25) +
       2 * exp(-(36/12)r3) +
       2 * exp(-(42/12) * ((9 / 10)r33333 + (1/ 10)x)) +
       2 * exp(-(48/12) * ((3 / 5)r33333 + (2 / 5)x)) +
       2 * exp(-(54/12) * ((3/ 10)r33333 + (7 / 10)x)) +
       102 * exp(-(60/12)x) - 103

n, r5 = newton(x0, f)

(4, 0.032118969698181545)

**Problem 9**: Given Treasury instruments:

$$
    \begin{matrix}
        - & \text{Coupon Rate} & \text{Price} \\
        6 - \text{Month T-bill} & 0 & 99.4565 \\
        12 - \text{Month T-bill} & 0 & 98.6196 \\
        2 - \text{Year T-note} & 2 & 101 \frac{17.5}{32} \\
        3 - \text{Year T-note} & 4.5 & 107 \frac{18}{32} \\
        5 - \text{Year T-note} & 3.125 & 102 \frac{8}{32} \\
        10 - \text{Year T-bond} & 4 & 103 \frac{8.5}{32}
    \end{matrix}
$$

The Treasury notes and bonds pay semiannual coupons. Assume continuously compounded interest.

(i) Obtain the zero rate curve from all the prices, except for the $3$ - Year T-note.

(ii) For the $3$-Year T-note, asked to conpute the zero rate curve and compute the relative pricing error.

**Solution**:

i) For the $6$-Month T-bill:

In [3]:
r05 = 2 * log(100 / 99.4565)

0.010899646693646998

For the $12$-Month T-bill:

In [4]:
r1 = 1 * log(100 / 98.6196)

0.013900161172756222

Next, I determine payment periods for the remaining coupons

In [5]:
get_periods(120, 6)

Stack{Int64}(Deque [[120, 114, 108, 102, 96, 90, 84, 78, 72, 66, 60, 54, 48, 42, 36, 30, 24, 18, 12, 6]])

then, setting $x = r(0, 2)$ and using linear interpolation for $t \in [1, 2]$,

$$
    r\left(0, \frac{18}{12}\right) = \frac{1}{2} \cdot r\left(0, 1\right) + \frac{1}{2}x.
$$

Finally, using bootstrapping and Newton's method for $r(0, 2)$:

In [6]:
x0 = 0.01
f(x) = 1 * exp((-6 / 12)r05) + 
       1 * exp((-1)r1) + 
       1 * exp((-18 / 12) * ((1 / 2)r1 + (1 / 2)x)) + 
       (100 + 1) * exp(-2x) - (101 + 17.5 / 32)

n, r2 = newton(x0, f)

(2, 0.012098547083637457)

In [7]:
r15 = (r1 + r2) / 2

0.01299935412819684

Now setting $x = r(0, 5)$ and using linear interpolation for $t \in [2, 5]$,

$$
    \begin{matrix}
        r\left(0, \frac{30}{12}\right) & = & \frac{5}{6} \cdot r(0, 2) + \frac{1}{6}x \\
        r\left(0, \frac{36}{12}\right) & = & \frac{2}{3} \cdot r(0, 2) + \frac{1}{3}x \\
        r\left(0, \frac{42}{12}\right) & = & \frac{1}{2} \cdot r(0, 2) + \frac{1}{2}x \\
        r\left(0, \frac{48}{12}\right) & = & \frac{1}{3} \cdot r(0, 2) + \frac{2}{3}x \\
        r\left(0, \frac{54}{12}\right) & = & \frac{1}{6} \cdot r(0, 2) + \frac{5}{6}x
    \end{matrix}
$$

Finally, using bootstrapping and Newton's method for $r(0, 5)$:

In [8]:
x0 = 0.01
f(x) = # known
       (3.125 / 2) * exp((-6 / 12)r05) +
       (3.125 / 2) * exp(-1r1) +
       (3.125 / 2) * exp((-18 / 12)r15) +
       (3.125 / 2) * exp((-2r2)) +
       # unkown
       (3.125 / 2) * exp((-30 / 12) * ((5 / 6)r2 + (1 / 6)x)) +
       (3.125 / 2) * exp((-36 / 12) * ((2 / 3)r2 + (1 / 3)x)) +
       (3.125 / 2) * exp((-42 / 12) * ((1 / 2)r2 + (1 / 2)x)) +
       (3.125 / 2) * exp((-48 / 12) * ((1 / 3)r2 + (2 / 3)x)) +
       (3.125 / 2) * exp((-54 / 12) * ((1 / 6)r2 + (5 / 6)x)) +
       (100 + (3.125 / 2)) * exp((-60 / 12)x) - (102 + (8 / 32))

n, r5 = newton(x0, f)

(3, 0.026823886324702104)

Finally, setting $x = r(0, 10)$ and using linear interpolation for $t \in [5, 10]$,

$$
    \begin{matrix}
        r\left(0, \frac{66}{12}\right) & = & \frac{9}{10} \cdot r(0, 5) + \frac{1}{10}x \\
        r\left(0, \frac{72}{12}\right) & = & \frac{4}{5} \cdot r(0, 5) + \frac{1}{5}x \\
        r\left(0, \frac{78}{12}\right) & = & \frac{7}{10} \cdot r(0, 5) + \frac{3}{10}x \\
        r\left(0, \frac{84}{12}\right) & = & \frac{3}{5} \cdot r(0, 5) + \frac{2}{5}x \\
        r\left(0, \frac{90}{12}\right) & = & \frac{1}{2} \cdot r(0, 5) + \frac{1}{2}x \\
        r\left(0, \frac{96}{12}\right) & = & \frac{2}{5} \cdot r(0, 5) + \frac{3}{5}x \\
        r\left(0, \frac{102}{12}\right) & = & \frac{3}{10} \cdot r(0, 5) + \frac{7}{10}x \\
        r\left(0, \frac{108}{12}\right) & = & \frac{1}{5} \cdot r(0, 5) + \frac{4}{5}x \\
        r\left(0, \frac{114}{12}\right) & = & \frac{1}{10} \cdot r(0, 5) + \frac{9}{10}x
    \end{matrix}
$$

Finally, using bootstrapping and Newton's method for $r(0, 10)$:

In [9]:
x0 = 0.01
f(x) = # known
       2 * exp((-6 / 12)r05) +
       2 * exp(-1r1) +
       2 * exp((-18 / 12)r15) +
       2 * exp((-2r2)) +
       2 * exp((-30 / 12) * ((5 / 6)r2 + (1 / 6)r5)) +
       2 * exp((-36 / 12) * ((2 / 3)r2 + (1 / 3)r5)) +
       2 * exp((-42 / 12) * ((1 / 2)r2 + (1 / 2)r5)) +
       2 * exp((-48 / 12) * ((1 / 3)r2 + (2 / 3)r5)) +
       2 * exp((-54 / 12) * ((1 / 6)r2 + (5 / 6)r5)) +
       2 * exp((-5r5)) +
       # unknown
       2 * exp((-66 / 12) * ((9 / 10)r5 + (1 / 10)x)) +
       2 * exp((-72 / 12) * ((4 / 5)r5 + (1 / 5)x)) +
       2 * exp((-78 / 12) * ((7 / 10)r5 + (3 / 10)x)) +
       2 * exp((-84 / 12) * ((3 / 5)r5 + (2 / 5)x)) +
       2 * exp((-90 / 12) * ((1 / 2)r5 + (1 / 2)x)) +
       2 * exp((-96 / 12) * ((2 / 5)r5 + (3 / 5)x)) +
       2 * exp((-102 / 12) * ((3 / 10)r5 + (7 / 10)x)) +
       2 * exp((-108 / 12) * ((1 / 5)r5 + (4 / 5)x)) +
       2 * exp((-114 / 12) * ((1 / 10)r5 + (9 / 10)x)) +
       (100 + 2) * exp((-120 / 12)x) - (103 + (8.5 / 32))

n, r10 = newton(x0, f)

(4, 0.03737166440440548)

ii) The zero rates computed for this example correspond to the following  zero rate curve which is piecewise linear between consecutive bond maturities:

$$
    r(0, t) = \begin{cases}
                    2(1 - t)\ r(0, 0.5) & + & (2t - 1)\ r(0, 1), \hspace{2em} & \text{if} & \frac{1}{2} \leq t \leq 1; \\
                    (2 - t)\ r(0, 1) & + & (t - 1)\ r(0, 2), \hspace{2em} & \text{if} & 1 \leq t \leq 2; \\
                    \frac{5 - t}{3}\ r(0, 2) & + & \frac{t - 2}{3}\ r(0, 5) & \text{if} & 2 \leq t \leq 5; \\
                    \frac{10 - t}{5}\ r(0, 5) & + & \frac{t - 5}{5}\ r(0, 10) & \text{if} & 5 \leq t \leq 10;
              \end{cases}
$$

With this zero rate curve, the value of the 3-year semiannual coupon bond with $4.5$ coupon rate is

In [20]:
B(x) = 2.25 * exp(-(1 / 2)r05) +
       2.25 * exp(-(1r1)) +
       2.25 * exp(-(1.5)r15) +
       2.25 * exp(-(2) * r2) +
       2.25 * exp(-(2.5) * (r2 + x) / 2) +
       102.25 * exp(-3x)

B (generic function with 1 method)

In [21]:
f(x) = # known
       2.25 * exp(-(1 / 2)r05) +
       2.25 * exp(-(1r1)) +
       2.25 * exp(-(1.5)r15) +
       2.25 * exp(-(2) * r2) +
       # unknown
       2.25 * exp(-(2.5) * (r2 + x) / 2) +
       102.25 * exp(-3x) - (107 + 18 / 32)

x0 = 0.01
n, r3 = newton(x0, f)

(3, 0.019157089237840604)

In [22]:
B(r3)

107.56250000000023

I have yet to figure out how they got the bond value to be $108.1930$, but anyway, following the formula

$$
    \% \text{error} = \frac{|\text{actual} - \text{predicted}|}{\text{actual}}
$$

the relative pricing error becomes

In [13]:
abs(107.5625 - 108.1930) / 107.5625

0.005861708309122583

I figured I can do linear interpolation to find what $r(0, 3)$ is for $t \in [2, 5]$. Then I get

$$
    r(0, 3) = \frac{2}{3}r(0, 2) + \frac{1}{3}r(0, 5)
$$

so then

In [24]:
r3 = (2 / 3)r2 + (1 / 3)r5

0.01700699349732567

In [ ]:
so then

In [25]:
B(r3)

108.19304171048005

In conclusion: I found an error in the following bond pricing formula (need to take **negative** values at the exponential curve):

```
B(x) = 2.25 * exp(**-**(1 / 2)r05) +
       2.25 * exp(**-**(1r1)) +
       2.25 * exp(**-**(1.5)r15) +
       2.25 * exp(-(2) * r2) +
       2.25 * exp(-(2.5) * (r2 + x) / 2) +
       102.25 * exp(-3x)
```
but that being corrected, I am getting the correct price, as per the book.

# Taylor's formula and Taylor series. ATM approximation of Black-Scholes formulas.

# $\S$ 6.2 Exercises

**Problem 11**: The $u$ and $d$ factors and risk-neutral probability $p$ of the price going up during one time step in the Cox-Ross-Rubinstein parametrization for a binomial tree are

$$
    u = A + \sqrt{A^2 - 1} \tag{6.2.1}
$$

$$
    d = A - \sqrt{A^2 - 1} \tag{6.2.2}
$$

$$
    p = \frac{e^{r\delta t} - d}{u - d} \tag{6.2.3} \hspace{5ex}
$$

where
$$
    A = \frac{1}{2}\ \left( e^{-r\delta t} + e^{(r + \sigma^2)\delta t} \right)
$$

Using Taylor expansions to show that, for a small time step $\delta t$, $u$ and $p$ may be approximated by

$$
    u = e^{\sigma\sqrt{\delta t}} \hspace{6em} \ \ \tag{6.2.4}
$$

$$
    d = e^{-\sigma\sqrt{\delta t}} \hspace{6em}\tag{6.2.5} \ \\
$$

$$
    p = \frac{1}{2} + \frac{1}{2}\ \left( \frac{r}{\sigma} - \frac{\sigma}{2} \right)\ \sqrt{\delta t}\tag{6.2.6}
$$

writing the Taylor expansions for $(6.2.1-6.2.3)$ and for $(6.2.4-6.2.6)$, showing that they are indentical if all the terms of order $O(\delta t)$ and smaller are neglected,

neglecting order $O((\delta t)^{3/2})$ and smaller, the solution to the problem is shown to be as follows (**worth reading pp. $160-162$ in literature**),

$$
    u = e^{\sigma\sqrt{\delta t}} + O\left((\delta t)^{3/2}\right)\hspace{6em}\hspace{2ex}
$$

$$
    d = e^{-\sigma\sqrt{\sigma t}} + O\left((\sigma t)^{3/2}\right)\hspace{6em}\hspace{1ex}
$$

$$
    p = \frac{1}{3} + \frac{1}{2\sigma}\left(r - \frac{\sigma^2}{2}\right)\sqrt{\sigma t} + O\left((\sigma t)^{3/2}\right)
$$

**Problem 12**:

(i)$\hspace{2ex}$Approximate value of ATM put option on a non-dividend-paying underlying asset with spot price $S = 60$, volaitility $\sigma = 0.25$, and maturity $T = 1$ year, if the constant risk is $r = 0$.

(ii)$\hspace{2ex}$Compute the Black-Scholes value $P_{BS,r=0,q=0}$ of the put option, and estimate the relative approximate error

$$
    \frac{|P_{BS,r=0.06,q=0.03} - P_{approx,r=0,q=0}|}{P_{BS,r=0,q=0}}\tag{6.2.7}
$$

(iii)$\hspace{2ex}$Assume that $r = 0.06$ and $q = 0.03$. Compute the approximate value $P_{approx,r=0.06,q=0.03}$ of an ATM put option and estimate the relative approximate error

$$
    \frac{|P_{BS,r=0.06,q=0.03} - P_{approx,r=0.06,q=0.03}|}{P_{BS,r=0.06,q=0.03}}\tag{6.2.8}
$$

where $P_{BS,r=0.06,q=0.03}$ is the Black-Scholes value of the put option.

One could answer (i) by using the approximate formula
$$
    P_{approx,r=0,q=0} = \sigma S\sqrt{\frac{T}{2\pi}}
$$
to obtain

In [25]:
D1 = Dict(
    "S" => 60,
    "K" => 60,
    "σ" => 0.25,
    "T" => 1.0,
    "r" => 0,
    "q" => 0,
    "t" => 0,
);

P_approx = D1["σ"] * D1["S"] * √(D1["T"] / (2π))

5.984134206021491

For (ii) it is easy to just plug in values to `black_scholes_approx`, to obtain

In [26]:
t = D1["t"]; S = D1["S"]; K = S; T = D1["T"]; σ = D1["σ"]; r = D1["r"]; q = r

P = black_scholes_approx(t, S, K, T, σ, r, q)[2][2]

5.96859189752395

and the relative approximate error is

In [27]:
abs(P - P_approx) / P

0.0026040159495555426

Finally, for (iii) the approximate formula

$$
    P_{approx,r\neq 0,q\neq 0} = \sigma S\sqrt{\frac{T}{2\pi}}\ \left(1 - \frac{(r + q)T}{2}\right) - \frac{(r - q)T}{2}S
$$

can be used along with `black_scholes_approx` to find the relative approximate error.

In [28]:
D1["r"] = 0.06; D1["q"] = 0.03; r = D1["r"]; q = D1["q"]
P_approx = σ * S * sqrt(T / (2 * π)) * (1 - (r + q)T / 2) - ((r - q)T / 2)S; P = black_scholes_approx(t, S, K, T, σ, r, q)[2][2]
abs(P - P_approx) / P

0.014761018843746894

$$
    \therefore \frac{|P_{BS,r=0.06,q=0.03} - P_{approx,r=0.06,q=0.03}|}{P_{BS,r=0.06,q=0.03}} \approxeq 0.014761 = 1.4761\%.
$$

**Problem 13**: It is interesting to note that the approximate formulas

$$
    C \approx \sigma S\sqrt{\frac{T}{2\pi}}\ \left(1 - \frac{(r + q)T}{2}\right) + \frac{(r - q)T}{2}S \\
    P \approx \sigma S\sqrt{\frac{T}{2\pi}}\ \left(1 - \frac{(r + q)T}{2}\right) - \frac{(r - q)T}{2}S
$$

for ATM call and put options do not satisfy the Put-Call parity:

$$
    P + Se^{qT} - C = S(e^{qT} - (r - q)T) \neq Se^{-rT} = Ke^{-rT}
$$

Based on the linear Taylor expansion $e^{-x} \approx 1 - x$, new approximation formulas for the price of ATM options which satisfy the Put-Call parity can be obtained by replacing $rT$ and $qT$ by $1 - e^{-rT}$ and $1 - e^{-qT}$, respectively. The resulting formulas are

$$
    C \approx \sigma S\sqrt{\frac{T}{2\pi}}\frac{e^{-qT} + e^{-rT}}{2} + \frac{S\ \left(e^{-qT} - e^{-rT}\right)}{2} \\
    P \approx \sigma S\sqrt{\frac{T}{2\pi}}\frac{e^{-qT} + e^{-rT}}{2} - \frac{S\ \left(e^{-qT} - e^{-rT}\right)}{2}
$$

The objective in this problem is to

1. show that the Put-Call parity is satisfied by the new approximation formulas.
2. for an ATM put with $S = 60$, $q = 0.03$, $\sigma = 0.25$, and $T = 1$, show how well the new approximation behaves by computing the relative approximate error.

For i:

It is easy to see that

$$
    P - C + Se^{-qT} = -2\frac{S\ (e^{-qT} - e^{-rT})}{2} + Se^{-qT} = Ke^{-rT}
$$

since K = S for ATM options.

For ii:

In [24]:
P_approx = D1["σ"]D1["S"] * √(D1["T"] / (2π)) * (exp(-q * T) + exp(-r * D1["T"])) / 2 - (D1["S"] * (exp(-q * T) - exp(-r * T))) / 2

4.8610308369861075

In [55]:
P = black_scholes_approx(t, S, K, T, σ, r, q)[2][2]
(P - P_approx) / P

0.0053108834720453565

Compared to the previous problem, $0.5311\%$ is a lower approximation error, hence this approximation is more accurate.

**Problem 14**: Asked to consider an ATM put option with strike $K = 40$ on a non-dividend paying asset with volatility $30\%$ and assume $r = 0$.

Asked to compute the relative approximation error of the approximation

$$
    P \approx \sigma S\sqrt{\frac{T}{2\pi}}
$$

if the put option expires in $1$, $3$, $5$, $10$, and $20$ years.

The precision of the approximation formula for ATM options are expected to decrease as the maturity of the option increases, which is indeed the case.

In [36]:
using DataFrames
D2 = Dict(
    "S" => 40,
    "K" => 40,
    "σ" => 0.30,
    "T" => [1, 3, 5, 10, 20],
    "r" => 0,
    "q" => 0,
    "t" => 0
);

t = D2["t"]; S = D2["S"]; K = S; σ = D2["σ"]; r = D2["r"]; q = D2["q"]; T = D2["T"];

P_approx = D2["σ"] * D2["S"] .* [√(T[I] / 2π) for I in 1:size(T)[1]]
P_BS = [black_scholes_approx(t, D2["S"], D2["K"], T[I], D2["σ"], r, q)[2][2] for I in 1:size(T)[1]]

approx_err = abs.(P_BS .- P_approx) ./ P_BS
df = DataFrames.DataFrame(T = [1, 3, 5, 10, 20], P_approx = P_approx, P_BS = P_BS, Approximation_Error = approx_err)

5×4 DataFrame
│ Row │ T     │ P_approx │ P_BS    │ Approximation_Error │
│     │ Int64 │ Float64  │ Float64 │ Float64             │
├─────┼───────┼──────────┼─────────┼─────────────────────┤
│ 1   │ 1     │ 4.78731  │ 4.76942 │ 0.00375107          │
│ 2   │ 3     │ 8.29186  │ 8.19951 │ 0.011263            │
│ 3   │ 5     │ 10.7047  │ 10.5074 │ 0.0187846           │
│ 4   │ 10    │ 15.1388  │ 14.5897 │ 0.0376324           │
│ 5   │ 20    │ 21.4095  │ 19.9066 │ 0.0754966           │

**Problem 15**: Asked to consider an ATM put option with strike $40$ on an asset with volatility $30\%$ and paying $2\%$ dividends continuously. Asked to assume that the interest rates are constant at $4.5\%$. Asked to compute the relative approximate error of the approximation

$$
    P_{approx1,r\neq 0, q\neq 0} = \sigma S\sqrt{\frac{T}{2\pi}}\ \left(1 - \frac{(r + q)T}{2}\right) - \frac{(r - q)T}{2}S
$$

although, I will take the exercise a step further by comparing to the (more accurate) approximation found in the previous problem,

$$
    P_{approx2,r\neq 0, q\neq 0} = \sigma S\sqrt{\frac{T}{2\pi}}\frac{e^{-qT} + e^{-rT}}{2} - \frac{S\ (e^{-qT} - e^{-rT})}{2} \hspace{0em}
$$

if the put option expires in $1$, $3$, $5$, $10$ and $20$ years. The approximate option values and the corresponding approximation error are given below:

In [94]:
D3 = Dict(
    "S" => 40,
    "K" => 40,
    "σ" => 0.30,
    "T" => [1, 3, 5, 10, 20],
    "r" => 0.045,
    "q" => 0.02,
    "t" => 0
);

S = D3["S"]; σ = D3["σ"]; T = D3["T"]; r = D3["r"]; q = D3["q"]; t = D3["t"]

P_approx1 = σ * S * sqrt.(T / (2π)) .* (1 .- (r + q)T / 2) .- ((r - q)T / 2)S
P_approx2 = σ * S * sqrt.(T / (2π)) .* (exp.(-q * T) + exp.(-r * T)) / 2 - S * (exp.(-q * T) .- exp.(-r * T)) ./ 2

P = [black_scholes_approx(t, S, K, T[I], σ, r, q)[2][2] for I in 1:size(T)[1]]; E1 = (P - P_approx1) ./ P; E2 = (P - P_approx2) ./ P

using DataFrames; df = DataFrames.DataFrame(P_approx1 = P_approx1, P_approx2 = P_approx2, P_BS = P, Error1 = E1, Error2 = E2)

5×5 DataFrame
│ Row │ P_approx1 │ P_approx2 │ P_BS    │ Error1     │ Error2       │
│     │ Float64   │ Float64   │ Float64 │ Float64    │ Float64      │
├─────┼───────────┼───────────┼─────────┼────────────┼──────────────┤
│ 1   │ 4.13172   │ 4.15056   │ 4.14914 │ 0.00419875 │ -0.000341789 │
│ 2   │ 5.9834    │ 6.16588   │ 6.15768 │ 0.0283027  │ -0.00133158  │
│ 3   │ 6.46522   │ 6.99056   │ 6.97141 │ 0.0726093  │ -0.00274636  │
│ 4   │ 5.21869   │ 7.40171   │ 7.33978 │ 0.288986   │ -0.00843698  │
│ 5   │ -2.50668  │ 6.25282   │ 6.05952 │ 1.41368    │ -0.0319003   │

**Problem 16**: A five year bond worth $101$ has duration $1.5$ years and convexity equal to $2.5$. Asked to use both the formula

$$
    \frac{\Delta B}{B} \approx\ -D\Delta y
$$

which does not include any convexity adjustment, and the formula

$$
    \frac{\Delta B}{B} \approx\ -D\Delta y + \frac{1}{2}C(\Delta y)^2
$$

to find the price of the bond if the yield increases by ten basis points ($10\ \text{BPS} \approx 0.001$), $50\ \text{BPS}$, $1\%$ and $2\%$, respectively.

Denoting by $B_{new,D}$ the approximate value given by the top formula for the value of the bond corresponding to the new yield, $\Delta B = B_{new,D} - B from which it follows that

$$
    B_{new,D} = B(1 - D\Delta y).
$$

Similarly, if $B_{new,D,C}$ is the approximate value for the value of the bond given by the second formula, it follows that

$$
    B_{new,D,C} = B(1 - D\Delta y + \frac{C}{2}(\Delta y)^2
$$

If $\Delta y \in \{0.001, 0.005, 0.01, 0.02\}$ as per the table below, where the last column is the approximate error, relative to the new value of the bond using duration alone, versus using duration and convexity, i.e.,

$$
    \frac{|B_{new,D,C} - B_{new,D}|}{B_{new,D,C}}
$$

In [107]:
D4 = Dict(
    "B" => 101,
    "D" => 1.5,
    "C" => 2.5,
    "Δy" => [0.001, 0.005, 0.01, 0.02],
);

B = D4["B"]; D = D4["D"]; C = D4["C"]; Δy = D4["Δy"]
BnewD = (1 .- D * Δy)B
BnewDC = (1 .- D * Δy .+ C * Δy .^ 2 / 2)B
Difference = abs.(BnewDC - BnewD) ./ BnewDC

using DataFrames; df = DataFrame(Δy = Δy, BnewD = BnewD, BnewDC = BnewDC, Difference = Difference)

4×4 DataFrame
│ Row │ Δy      │ BnewD   │ BnewDC  │ Difference  │
│     │ Float64 │ Float64 │ Float64 │ Float64     │
├─────┼─────────┼─────────┼─────────┼─────────────┤
│ 1   │ 0.001   │ 100.849 │ 100.849 │ 1.25188e-6  │
│ 2   │ 0.005   │ 100.243 │ 100.246 │ 3.14852e-5  │
│ 3   │ 0.01    │ 99.485  │ 99.4976 │ 0.000126887 │
│ 4   │ 0.02    │ 97.97   │ 98.0205 │ 0.000515198 │

**Problem 17**: (this is a **parallel shift** and **portfolio immunization** problem) asks to consider a bond portfolio worth $\$50\text{mil}$ with DV$01$ equal to $\$10,000$ and dollar convexity equal to $\$400\text{mil}$.

I. Assume that the yield curve moves up by fifty basis points. What is the new value of the bond portfolio?

II. The following bonds are available for trading:

In [7]:
txb = ["Bond 1", "Bond 2"]
prc = [1000, 100]
val = [1100, 106]
dur = [2.5, 3]
cvx = [12, 9]

using DataFrames; df = DataFrame(_ = txb, Principal = prc, Value = val, Duration = dur, Convexity = cvx)

2×5 DataFrame
│ Row │ _      │ Principal │ Value │ Duration │ Convexity │
│     │ String │ Int64     │ Int64 │ Float64  │ Int64     │
├─────┼────────┼───────────┼───────┼──────────┼───────────┤
│ 1   │ Bond 1 │ 1000      │ 1100  │ 2.5      │ 12        │
│ 2   │ Bond 2 │ 100       │ 106   │ 3.0      │ 9         │

How can the portfolio be immunized?

I. For one basis point shift upward of the yield curve, the value of the bond portfolio decreases by an amount equal to its $\text{DV}01$, so $\$10,000.$ Thus a fifty basis points upward move of the yield curve will result in a loss of

$$
    50 \cdot \$10,000 = \$500,000
$$

in the bond portfolio. The new value of the portfolio will be approximately $\$50\text{mil} - \$500,000 = \$49.5\text{mil}$.

A more precise approach includes a convexity adjustment. Asked to recall that, if the yield curve experiences a small **parallel shift** of size \delta r, the change \Delta V in the value of the bond portfolio can be approximated as

$$
    \Delta V \approx -D_{\$}(V)\delta r + \frac{C_{\$}(V)}{2}(\delta r)^2\tag{6.2.9}
$$

where $D_{\$}(V)$ and $C_{\$}(V)$ are the dollar duration and dollar convexity of the portfolio, respectively.

The $\text{DV}01$ of the bond portfolio is equal to the dollar duration of the bond portfolio divided by $10,000$, i.e.,

$$
    \text{DV}01(V) = \frac{D_{\$}(V)}{10-000}
$$

Thus, $D_{\$}(V) = \text{DV}01(V) \cdot 10-000 = \$100\text{mil}$.

Using $(6.2.9)$ with $\delta r = 50\text{bps} =  0.005$, $D_\$(V) = \$100\text{mil}$, and $C_\$(V) = \$400\text{mil}$, the change in the value of the bond portfolio becomes

$$
    \Delta V \approx -$100\text{mil}\cdot 0.005 + \frac{\$400\text{mil}}{2}\cdot(0.005)^2 \\
    = -$500-000 + $5-000 \hspace{3em}\hspace{1ex} \\
    = -$495-000.\hspace{7em}\hspace{1ex}
$$

The new value of the bond portfolio is $V_{new} = V_{old} + \Delta V \approx \$50\text{mil} - \$495-000 = \$49-505-000$. (**View discussion**).

II. Asked to denote by $B_1$ and $B_2$ the sizes of the positions taken in the two bonds available for trading. The value of the new portfolio is
    
$$
    \Pi = V + B_1 + B_2.
$$

Dollar duration and convexity of bond positions:

$$
    D_\$(B_1) = B_1D_1 = 2.5B_1; C_\$(B_1) = B_1C_1 = 12B_1 \\
    D_\$(B_2) = B_2D_2 = 3.0B_2; C_\$(B_2) = B_2C_2 = 09B_2
$$

Asked to recall that $D_\$(V) = \$100\text{mil}$ and $C_\$(V) = \$400\text{mil}$. The dollar duration and the dollar convexity of the new portfolio are

$$
    D_\$(\Pi) = D_\$(V) + D_\$(B_1) + D_\$(B_2) = \$100\text{mil} + 02.5B_1 + 3B_2 \\
    C_\$(\Pi) = C_\$(V) + C_\$(B_1) + C_\$(B_2) = \$400\text{mil} + 12.0B_1 + 9B_2
$$

What is left is solving a linear system for $D_\$(\Pi) = 0$ and $C_\$(\Pi) = 0$. That linear system is

$$
    \begin{cases}
        D_\$(V) = 02.5B_1 + 3B_2 = -\$100\text{mil} \\
        C_\$(V) = 12.0B_1 + 9B_2 = -\$400\text{mil}
    \end{cases}
$$

Below, initial values $B_0 = [1\text{mil}, 1\text{mil}]$ were taken for an initial guess of the prices for $B_1$ and $B_2$.

In [6]:
B0 = [1, 1]
f(B) = 2.5 * B[1] + 3 * B[2] + 100
g(B) = 12 * B[1] + 9 * B[2] + 400
F = Function[]; push!(F, f); push!(F, g)

n, x = approxNewton(B0, F)

(2, [-22.222222222222218, -14.81481481481482])

This was accurate to eight significant figures, as in the book (scale is in the millions). These are the sizes of the bond positions that must be added to the original portfolio in order to immunize it from small changes in the yield curve.

$\therefore$ in order to immunize the given portfolio, one would take the following short investment positions

In [12]:
abs.(x ./ val * 1000000)

2-element Array{Float64,1}:
  20202.020202020198
 139762.40391334734

units of Bond $1$ and Bond $2$, respectively.

**Problem 18**:

I. If the current zero rate curve is

$$
    r_1(0, t) = 0.025 + \frac{1}{100}\exp\left({-\frac{t}{100}}\right) + \frac{t}{100(t + 1)},
$$

asked to find the yield of the four year semiannual coupon bond with a coupon rate 6\%. Asked to assume that the interest is compounded continuously and that the face value of the bond is 100.

II. If the zero rate have a parallel shift up by $10$, $20$, $50$, $100$ and $200$ basis points, respectively, i.e., if the zero rate curve changes from $r_1(0,t)$ to $r_2(0, t) = r_1(0, t) + dr$, with $dr$ as follows,

In [13]:
BP = 0.0001
dr = [10, 20, 50, 100, 200] * BP

5-element Array{Float64,1}:
 0.001
 0.002
 0.005
 0.01
 0.02

by how much does the yield of the bond increase in each case?

**Solution**

I. The bond provides coupon payments equal to three every six months until $3.5$ years from now, and a final cash flow of $103$ in four years. Below is an attempt to discount the cash flows to the present using the zero rate curve $r_1(0,t)$ and `get_cash_flow` as well as `bond_price_zero_rate` from `a2x.jl`, arriving at a solution for the bond price, $B$.

In [126]:
Y = 4; n = convert(Int64, Y * 12 / 6); m = 2; T = 12 * 4; F = 100; C = 0.06; t_cash_flow, v_cash_flow = get_cash_flow(n, m, T, F, C)

f(x;z = 0) = 0.025 + (1 / F)exp(-x / F) + x / (F * (x + 1)) + z; disc, B = bond_price_zero_rate(n, t_cash_flow, v_cash_flow, f); B

106.19952411534813

Fortunately, for the bond yield, there is a function, `bond_yield`, which was used for an earlier problem. Plugging in the values would return

In [127]:
y = bond_yield(B, n, t_cash_flow, v_cash_flow)

0.04251072909468839

or $4.2511\%$.

II. Small **parallel shifts** in the zero rate curve of `[0.001, 0.002, 0.005, 0.01, 0.02]` would give the following new bond prices and bond yields.

In [174]:
disc = zeros(40); disc = reshape(disc, 8, 5); B = zeros(5)

[(disc[:, I], B[I]) = bond_price_zero_rate(n, t_cash_flow, v_cash_flow, f, z = dr[I]) for I in 1:size(dr)[1]]

y_new = [bond_yield(B[I], n, t_cash_flow, v_cash_flow) for I in 1:size(dr)[1]]

using DataFrames; df = DataFrame(zero_rate_shift = ["10bps = 0.001", "20bps = 0.002", "50bps = 0.005", "100bps = 0.01", "200bps = 0.02"],
                                 new_bond_price = B, new_yield = y_new, yield_increase = y_new .- y)

5×4 DataFrame
│ Row │ zero_rate_shift │ new_bond_price │ new_yield │ yield_increase │
│     │ String          │ Float64        │ Float64   │ Float64        │
├─────┼─────────────────┼────────────────┼───────────┼────────────────┤
│ 1   │ 10bps = 0.001   │ 105.815        │ 0.0435105 │ 0.000999786    │
│ 2   │ 20bps = 0.002   │ 105.432        │ 0.0445103 │ 0.00199957     │
│ 3   │ 50bps = 0.005   │ 104.292        │ 0.0475097 │ 0.00499893     │
│ 4   │ 100bps = 0.01   │ 102.42         │ 0.0525086 │ 0.00999784     │
│ 5   │ 200bps = 0.02   │ 98.7829        │ 0.0625064 │ 0.0199956      │

As expected, the increase of the yield of the bond is slightly smaller, but very close to, the parallel shift of the zero rate curve, i.e., $y_2 - y \approx dr$.